<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Get started • SEMgraph</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get started">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SEMgraph</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/SEMgraph_tutorial.html">Get started</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fernandoPalluzzi/SEMgraph/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get started</h1>
                        <h4 data-toc-skip class="author">Fernando
Paluzzi, Mario Grassi</h4>
            
            <h4 data-toc-skip class="date">2025-12-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/fernandoPalluzzi/SEMgraph/blob/HEAD/vignettes/SEMgraph_tutorial.Rmd" class="external-link"><code>vignettes/SEMgraph_tutorial.Rmd</code></a></small>
      <div class="d-none name"><code>SEMgraph_tutorial.Rmd</code></div>
    </div>

    
    
<p>The following section offers an overview of <strong>SEMgraph</strong>
functionalities. Starting from model fitting, it will gently introduce
functions for model learning, weighting, clustering, and evaluation of
causal effects and model perturbation. This section includes:</p>
<ol style="list-style-type: decimal">
<li><strong>Supplementary packages installation</strong></li>
<li><strong>Causal effects estimation, model learning, extension, and
clusterinng (Amyotrophic Lateral Sclerosis dataset)</strong></li>
<li><strong>Gene Set Analysis and perturbed subnetwork/module extraction
(Frontotemporal Dementia dataset)</strong></li>
</ol>
<p> </p>
<div class="section level2">
<h2 id="supplementary-packages-">1. Supplementary packages.<a class="anchor" aria-label="anchor" href="#supplementary-packages-"></a>
</h2>
<p>Besides the required packages, SEMgraph suggests the use of the R
package <strong>org.Hs.eg.db</strong> in the Bioconductor project, for
gene ID conversion. SEMgraph uses entrez IDs to avoid special chatacters
(such as hyphens or slashes), but it can use official gene symbols as
labels.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span></code></pre></div>
<p>The package <a href="https://github.com/fernandoPalluzzi/SEMdata" class="external-link"><strong>SEMdata</strong></a>
contains useful high-throughput sequencing data, reference networks, and
pathways for SEMgraph training:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#devtools::install_github("fernandoPalluzzi/SEMdata")</span></span></code></pre></div>
<p> </p>
</div>
<div class="section level2">
<h2 id="amyotrophic-lateral-sclerosis-als-data-analysis-">2. Amyotrophic Lateral Sclerosis (ALS) data analysis.<a class="anchor" aria-label="anchor" href="#amyotrophic-lateral-sclerosis-als-data-analysis-"></a>
</h2>
<div class="section level3">
<h3 id="the-als-dataset-">2.1. The ALS dataset.<a class="anchor" aria-label="anchor" href="#the-als-dataset-"></a>
</h3>
<p>Load the necessary libraries for run the tutorial:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fernandoPalluzzi/SEMgraph" class="external-link">SEMgraph</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SEMdata</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre></div>
<p><strong>SEMdata</strong> provides the ALS RNA-seq dataset of 139
cases and 21 healthy controls, from Tam O.H. <em>et al.</em>, 2019 (GEO
accession: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE124439" class="external-link">GSE124439</a>).
Raw data were pre-processed applying batch effect correction to remove
data production center and brain area biases. Using multidimensional
scaling-based clustering, ALS-specific and HC-specific clusters were
generated. Misclassified samples were blacklisted and removed from the
dataset. Since the expression of many genes is significantly different
from Gaussian, we apply a nonparanormal transform to relax the normality
assumption.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ALS input graph</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH 5d44052 DNW- 32 47 -- </span></span>
<span><span class="co">## + attr: name (v/c), weight (e/n)</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Label graph: Convert Entrez identifiers to gene symbols</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>,</span>
<span>                                 column <span class="op">=</span> <span class="st">'SYMBOL'</span>,</span>
<span>                                 keytype <span class="op">=</span> <span class="st">'ENTREZID'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ALS RNA-seq expression data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">exprs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]   160 17695</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># group = {1: case, 0: control} vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##   0   1 </span></span>
<span><span class="co">##  21 139</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Nonparanormal transform</span></span>
<span><span class="va">data.npn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transformData.html">transformData</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">exprs</span><span class="op">)</span><span class="op">$</span><span class="va">data</span></span></code></pre></div>
<pre><code><span><span class="co">## Conducting the nonparanormal transformation via shrunkun ECDF...done.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="model-fitting-">2.2. Model fitting.<a class="anchor" aria-label="anchor" href="#model-fitting-"></a>
</h3>
<p><strong>SEMgraph</strong> offers three main modes of model fitting:
(i) common model fitting, (ii) node perturbation evaluation, and (iii)
edge perturbation evaluation. <strong>SEMgraph</strong> will
automatically take care of applying shrinkage methods in case of high
dimensionality (#variables &gt;&gt; #subjects), heuristics and
parallelization settings for large graphs. In addition, perturbation
evaluation enables the extraction of differentially regulated nodes
(DRNs) and edges (DREs).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ALS model fitting (sem0)</span></span>
<span><span class="co"># The whole dataset is used to fit the model and perturbation is not evaluated (group = NULL)</span></span>
<span><span class="va">sem0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, data <span class="op">=</span> <span class="va">data.npn</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NLMINB solver ended normally after 1 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 10.92504  srmr: 0.2858859</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># verbose output</span></span>
<span><span class="co">#summary(sem0$fit)</span></span>
<span><span class="co"># short output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">sem0</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     lhs op  rhs    est    se      z pvalue ci.lower ci.upper</span></span>
<span><span class="co">## 1 10452  ~ 6647  0.037 0.079  0.466  0.641   -0.118    0.192</span></span>
<span><span class="co">## 2  1432  ~ 5606  0.397 0.082  4.839  0.000    0.236    0.557</span></span>
<span><span class="co">## 3  1432  ~ 5608  0.578 0.082  7.047  0.000    0.417    0.738</span></span>
<span><span class="co">## 4  1616  ~ 7132  0.245 0.110  2.236  0.025    0.030    0.461</span></span>
<span><span class="co">## 5  1616  ~ 7133 -0.036 0.110 -0.324  0.746   -0.251    0.180</span></span>
<span><span class="co">## 6  4217  ~ 1616 -0.074 0.079 -0.943  0.346   -0.229    0.080</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ALS model fitting (sem1)</span></span>
<span><span class="co"># Common model and group influence on nodes (node perturbation)</span></span>
<span></span>
<span><span class="va">sem1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, data <span class="op">=</span> <span class="va">data.npn</span>, group <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NLMINB solver ended normally after 6 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 11.02582  srmr: 0.274885 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node activation: 7.104523e-08 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node inhibition: 0.01068785</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># verbose output</span></span>
<span><span class="co">#summary(sem1$fit)</span></span>
<span><span class="co"># short output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">sem1</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     lhs op   rhs    est    se      z pvalue ci.lower ci.upper</span></span>
<span><span class="co">## 1 10452  ~ group -0.150 0.078 -1.913  0.056   -0.303    0.004</span></span>
<span><span class="co">## 2  1432  ~ group -0.042 0.073 -0.578  0.563   -0.186    0.101</span></span>
<span><span class="co">## 3  1616  ~ group  0.025 0.080  0.314  0.754   -0.131    0.181</span></span>
<span><span class="co">## 4   317  ~ group  0.218 0.077  2.832  0.005    0.067    0.370</span></span>
<span><span class="co">## 5  4217  ~ group  0.176 0.078  2.273  0.023    0.024    0.328</span></span>
<span><span class="co">## 6  4741  ~ group  0.343 0.075  4.547  0.000    0.195    0.490</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Colored group effects on nodes and direct effects adjusted for group</span></span>
<span><span class="co">#(red/pink: activation, blue/lightblue: repression)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">sem1</span><span class="op">$</span><span class="va">graph</span>, main<span class="op">=</span><span class="st">"Colored ALS graph"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-3-1.png" style="width:95.0%"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ALS model fitting (sem2)</span></span>
<span><span class="co"># One model for each group and group influence on edges (edge perturbation)</span></span>
<span></span>
<span><span class="va">sem2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, <span class="va">data.npn</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>, fit <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Estimating optimal shrinkage intensity lambda (correlation matrix): 0.4313 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## NLMINB solver ended normally after 7 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 16.08203  srmr: 0.2971455 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of edge activation: 0.004597565 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of edge inhibition: 0.9534567</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># verbose output</span></span>
<span><span class="co">#summary(sem2$fit)</span></span>
<span><span class="co"># short output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">sem2</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     lhs op  rhs    est    se      z pvalue ci.lower ci.upper group</span></span>
<span><span class="co">## 1 10452  ~ 6647  0.198 0.214  0.928  0.353   -0.221    0.618     1</span></span>
<span><span class="co">## 2  1432  ~ 5606  0.101 0.209  0.482  0.630   -0.309    0.511     1</span></span>
<span><span class="co">## 3  1432  ~ 5608  0.346 0.209  1.654  0.098   -0.064    0.757     1</span></span>
<span><span class="co">## 4  1616  ~ 7132  0.120 0.240  0.499  0.618   -0.350    0.590     1</span></span>
<span><span class="co">## 5  1616  ~ 7133  0.118 0.240  0.493  0.622   -0.352    0.588     1</span></span>
<span><span class="co">## 6  4217  ~ 1616 -0.097 0.217 -0.445  0.656   -0.522    0.329     1</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perturbed graph elements</span></span>
<span></span>
<span><span class="co"># Differentially Regualted Nodes (DRNs)</span></span>
<span><span class="va">DRN</span> <span class="op">&lt;-</span> <span class="va">sem1</span><span class="op">$</span><span class="va">gest</span><span class="op">[</span><span class="va">sem1</span><span class="op">$</span><span class="va">gest</span><span class="op">$</span><span class="va">pvalue</span> <span class="op">&lt;</span> <span class="fl">0.05</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">DRN</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 16</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DRN</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      lhs op   rhs   est    se     z pvalue ci.lower ci.upper</span></span>
<span><span class="co">## 4    317  ~ group 0.218 0.077 2.832  0.005    0.067    0.370</span></span>
<span><span class="co">## 5   4217  ~ group 0.176 0.078 2.273  0.023    0.024    0.328</span></span>
<span><span class="co">## 6   4741  ~ group 0.343 0.075 4.547  0.000    0.195    0.490</span></span>
<span><span class="co">## 8   4747  ~ group 0.223 0.062 3.620  0.000    0.102    0.344</span></span>
<span><span class="co">## 9  54205  ~ group 0.188 0.069 2.730  0.006    0.053    0.322</span></span>
<span><span class="co">## 10  5530  ~ group 0.160 0.072 2.224  0.026    0.019    0.301</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Differentially Regulated Edges (DREs)</span></span>
<span><span class="va">DRE</span> <span class="op">&lt;-</span> <span class="va">sem2</span><span class="op">$</span><span class="va">dest</span><span class="op">[</span><span class="va">sem2</span><span class="op">$</span><span class="va">dest</span><span class="op">$</span><span class="va">pvalue</span> <span class="op">&lt;</span> <span class="fl">0.05</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">DRE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DRE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     lhs op  rhs d_est  d_se   d_z pvalue d_lower d_upper</span></span>
<span><span class="co">## 28 5532  ~ 6647 0.449 0.227 1.982  0.047   0.005   0.893</span></span>
<span><span class="co">## 30 5534  ~ 6647 0.584 0.229 2.546  0.011   0.135   1.034</span></span>
<span><span class="co">## 34 5603  ~ 5606 0.496 0.239 2.073  0.038   0.027   0.965</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Residual Iterative Conditional Fitting (RICF) </span></span>
<span><span class="co"># Fast node perturbation fitting for large graphs (SE estimation disabled)</span></span>
<span></span>
<span><span class="va">ricf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, <span class="va">data.npn</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>, algo <span class="op">=</span> <span class="st">"ricf"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## RICF solver ended normally after 2 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 11.02558  srmr: 0.2747457 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node activation: 7.074471e-08 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node inhibition: 0.009522031</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">ricf1</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   lhs op   rhs    est</span></span>
<span><span class="co">## 1 317  ~ group  0.218</span></span>
<span><span class="co">## 2 572  ~ group -0.106</span></span>
<span><span class="co">## 3 581  ~ group -0.161</span></span>
<span><span class="co">## 4 596  ~ group  0.157</span></span>
<span><span class="co">## 5 598  ~ group  0.107</span></span>
<span><span class="co">## 6 836  ~ group  0.245</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Other possible output</span></span>
<span><span class="co">#summary(ricf1$fit)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ricf1</span><span class="op">$</span><span class="va">gest</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Test    Stat tail       pvalue</span></span>
<span><span class="co">## 317    t  2.8143   &gt;&lt; 0.0073759375</span></span>
<span><span class="co">## 572    t -1.3373   &gt;&lt; 0.1889142563</span></span>
<span><span class="co">## 581    t -2.0472   &gt;&lt; 0.0349464466</span></span>
<span><span class="co">## 596    t  2.0117   &gt;&lt; 0.0549088354</span></span>
<span><span class="co">## 598    t  1.3567   &gt;&lt; 0.1708148899</span></span>
<span><span class="co">## 836    t  3.6671   &gt;&lt; 0.0004791965</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Constrained Gaussian Graphical Modeling (CGGM)</span></span>
<span><span class="co"># Fast edge perturbation fitting for large graphs</span></span>
<span><span class="va">cggm2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, <span class="va">data.npn</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>, fit <span class="op">=</span> <span class="fl">2</span>, algo <span class="op">=</span> <span class="st">"cggm"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GGM (de-biased nodewise L1) solver ended normally after 0 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 16.11538  srmr: 0.2977598 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of edge activation: 8.441818e-05 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of edge inhibition: 0.7365028</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">cggm2</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     lhs op  rhs    est    se z_test pvalue ci.lower ci.uppper group</span></span>
<span><span class="co">## 1 10452  ~ 6647  0.349 0.204  1.707  0.088   -0.052     0.750     1</span></span>
<span><span class="co">## 2 84134  ~ 6647  0.607 0.173  3.497  0.000    0.267     0.947     1</span></span>
<span><span class="co">## 3 54205  ~  581 -0.584 0.175 -3.340  0.001   -0.927    -0.242     1</span></span>
<span><span class="co">## 4 54205  ~  572  0.272 0.177  1.534  0.125   -0.076     0.620     1</span></span>
<span><span class="co">## 5 54205  ~  596  0.308 0.197  1.567  0.117   -0.077     0.694     1</span></span>
<span><span class="co">## 6 54205  ~  598  0.283 0.197  1.435  0.151   -0.103     0.669     1</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Other possible output</span></span>
<span><span class="co">#summary(cggm2$fit$Group_0)</span></span>
<span><span class="co">#summary(cggm2$fit$Group_1)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cggm2</span><span class="op">$</span><span class="va">dest</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     lhs op  rhs  d_est  d_se    d_z pvalue d_lower d_upper</span></span>
<span><span class="co">## 1 10452  ~ 6647 -0.363 0.221 -1.638  0.101  -0.797   0.071</span></span>
<span><span class="co">## 2 84134  ~ 6647 -0.443 0.193 -2.301  0.021  -0.821  -0.066</span></span>
<span><span class="co">## 3 54205  ~  581  0.146 0.194  0.751  0.453  -0.235   0.527</span></span>
<span><span class="co">## 4 54205  ~  572 -0.273 0.194 -1.404  0.160  -0.654   0.108</span></span>
<span><span class="co">## 5 54205  ~  596 -0.512 0.209 -2.449  0.014  -0.923  -0.102</span></span>
<span><span class="co">## 6 54205  ~  598 -0.381 0.217 -1.754  0.079  -0.806   0.045</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="total-effect-estimation-as-average-causal-effect-ace-">2.3. Total effect estimation as Average Causal Effect (ACE).<a class="anchor" aria-label="anchor" href="#total-effect-estimation-as-average-causal-effect-ace-"></a>
</h3>
<p>Suppose having a path P = X -&gt; M1 -&gt; … -&gt; Mk -&gt; Y between
two nodes X and Y, connected through k mediators. The total effect (TE =
DE + IE) is the sum of the direct effect X -&gt; Y, DE = b(X,Y), and the
indirect effect through the mediators, IE = b(X,M1)*b(M1,M2)*…*b(Mk,Y).
One convenient way of estimating the TE is through the definition of ACE
by <a href="https://doi.org/10.1177/0049124198027002004" class="external-link">Pearl J,
1998</a>. The simplest estimation of the TE as ACE is possible in
directed acyclic graphs (DAGs), thorugh linear regression. The parent
set pa(X) of X blocks all backdoor (i.e., confounding) paths from X to
Y, and the ACE is equal to the coefficient b(X,Y) in a multiple
regression Y ~ X + pa(X).</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Average Causal Effect (ACE)</span></span>
<span><span class="co"># optimal adjustement set (type = "optimal")</span></span>
<span><span class="co"># source -&gt; sink ACE (effect = "source2sink")</span></span>
<span><span class="co"># Benjamini-Hochberg multiple test adjustment (method = "BH")</span></span>
<span><span class="co"># 5% significance level (alpha = 0.05)</span></span>
<span></span>
<span><span class="va">ace</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMace.html">SEMace</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, data <span class="op">=</span> <span class="va">data.npn</span>,</span>
<span>              type <span class="op">=</span> <span class="st">"optimal"</span>, effect <span class="op">=</span> <span class="st">"source2sink"</span>,</span>
<span>              method <span class="op">=</span> <span class="st">"BH"</span>, alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Frequency distribution of path length from X to Y :</span></span>
<span><span class="co">##  1  2  3  4  5 </span></span>
<span><span class="co">## 12  1  3  1  6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  ACE= 1 of 23 ACE= 2 of 23 ACE= 3 of 23 ACE= 4 of 23 ACE= 5 of 23 ACE= 6 of 23 ACE= 7 of 23 ACE= 8 of 23 ACE= 9 of 23 ACE= 10 of 23 ACE= 11 of 23 ACE= 12 of 23 ACE= 13 of 23 ACE= 14 of 23 ACE= 15 of 23 ACE= 16 of 23 ACE= 17 of 23 ACE= 18 of 23 ACE= 19 of 23 ACE= 20 of 23 ACE= 21 of 23 ACE= 22 of 23 ACE= 23 of 23</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Sort by decreasing abs(z) value</span></span>
<span><span class="va">ace</span> <span class="op">&lt;-</span> <span class="va">ace</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">ace</span><span class="op">$</span><span class="va">z</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ace</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ace</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       pathL  sink op source    est    se      z pvalue ci.lower ci.upper</span></span>
<span><span class="co">## 66471     1  4747 &lt;-   6647  0.514 0.063  8.113      0    0.390    0.639</span></span>
<span><span class="co">## 317       2   836 &lt;-    317  0.472 0.061  7.737      0    0.352    0.592</span></span>
<span><span class="co">## 11        1 79139 &lt;-   6647  0.522 0.068  7.723      0    0.390    0.655</span></span>
<span><span class="co">## 13        1  5532 &lt;-   6647  0.521 0.068  7.700      0    0.389    0.654</span></span>
<span><span class="co">## 16        1  5535 &lt;-   6647 -0.462 0.070 -6.565      0   -0.600   -0.324</span></span>
<span><span class="co">## 6647      4   836 &lt;-   6647  0.430 0.067  6.433      0    0.299    0.561</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SOD1-CASP3 path extraction and fitting</span></span>
<span></span>
<span><span class="va">source</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">ace</span><span class="op">$</span><span class="va">source</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sink</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">ace</span><span class="op">$</span><span class="va">sink</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMpath.html">SEMpath</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, <span class="va">data.npn</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>,</span>
<span>                from <span class="op">=</span> <span class="va">source</span>, to <span class="op">=</span> <span class="va">sink</span>,</span>
<span>                path <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>                verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Path: 6647 -&gt; 836 size- 5 4 --</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## NLMINB solver ended normally after 1 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 24.52913  srmr: 0.2068706 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node activation: 4.363726e-06 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node inhibition: 0.9267063</span></span></code></pre>
<p><img src="figure-html/unnamed-chunk-4-1.png" style="width:95.0%"></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract all directed paths, fit them, and evaluate their perturbation:</span></span>
<span></span>
<span><span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pathFinder.html">pathFinder</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, <span class="va">data.npn</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>, ace <span class="op">=</span> <span class="va">ace</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  ACE= 1 of 10 ACE= 2 of 10 ACE= 3 of 10 ACE= 4 of 10 ACE= 5 of 10 ACE= 6 of 10 ACE= 7 of 10 ACE= 8 of 10 ACE= 9 of 10 ACE= 10 of 10</span></span>
<span><span class="co">## Found 3 significant ACEs with &gt; 2 nodes</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">paths</span><span class="op">$</span><span class="va">dfp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   pathId sink op source n.nodes n.edges dev_df  srmr V.pv.act V.pv.inh</span></span>
<span><span class="co">## 1   P317  836 &lt;-    317       3       2 21.328 0.091 0.000042 0.232830</span></span>
<span><span class="co">## 2  P6647  836 &lt;-   6647       5       4 24.526 0.207 0.000004 0.926706</span></span>
<span><span class="co">## 3   P581  836 &lt;-    581       4       3  3.699 0.046 0.000959 0.179103</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="model-estimation-strategies-">2.4. Model estimation strategies.<a class="anchor" aria-label="anchor" href="#model-estimation-strategies-"></a>
</h3>
<p>The input graph is a picture of the current knowledge. Besides the
evaluation of known relationships, data can be used to infer new ones.
Four model search strategies are implemented in the
<code>modelSearch</code> function. The <em>basic</em> one is completely
data-driven and requires an input graph only to establish the initial
topological order. In the <em>direct</em> strategy, The input graph
structure is improved through direct (i.e., adjacent) link search,
followed by reference-based (“gnet” argument) interaction validation and
import from the reference network, with no mediators (d = 1). In the
<em>outer</em> strategy, new interactions and connectors (i.e.,
mediators) will be searched and imported from the reference network.
This strategy is analogous to the “outer” one, but disables external
mediator search. In other words, new indirect paths are generated by
adding new interactions between nodes only from the input model.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Model Search with search = "basic" and beta = 0.1</span></span>
<span><span class="co"># The beta argument determines the minimum absolute LASSO coefficient</span></span>
<span><span class="co"># for an edge to be included in the output model.</span></span>
<span><span class="co"># Reducing beta (up to 0) will increase model complexity.</span></span>
<span></span>
<span><span class="co"># This strategy is data-driven (gnet = NULL) and </span></span>
<span><span class="co"># only direct connections are added (d = 0; i.e., no new mediators),</span></span>
<span><span class="co"># these edges are colored in green</span></span>
<span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelSearch.html">modelSearch</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>,</span>
<span>                     data <span class="op">=</span> <span class="va">data.npn</span>,</span>
<span>                     gnet <span class="op">=</span> <span class="cn">NULL</span>, d <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                     search <span class="op">=</span> <span class="st">"basic"</span>,</span>
<span>                     beta <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                     alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                     verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Step1: BAP deconfounding...</span></span>
<span><span class="co">## Step2: DAG estimation...</span></span>
<span><span class="co">## Step3: DAG resize (remove edges/add nodes)...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## None DAG resize for basic search ! </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Done.</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert Entrez identifiers to gene symbols</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>,</span>
<span>                              column <span class="op">=</span> <span class="st">'SYMBOL'</span>,</span>
<span>                              keytype <span class="op">=</span> <span class="st">'ENTREZID'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Colored plot: orange/cyan: sink/source nodes, green/gray: new/old edges</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">graph</span>, main<span class="op">=</span><span class="st">"ALS model"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-5-1.png" style="width:95.0%"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Node perturbation of the ALS model</span></span>
<span><span class="va">pert</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">graph</span>, <span class="va">model</span><span class="op">$</span><span class="va">data</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NLMINB solver ended normally after 7 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 1.859277  srmr: 0.0818185 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node activation: 4.195065e-05 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node inhibition: 0.002192286</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Colored plot: red/pink: activation, blue/lightblue: repression</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">pert</span><span class="op">$</span><span class="va">graph</span>, main<span class="op">=</span><span class="st">"ALS model"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-5-2.png" style="width:95.0%"></p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SOD1-NEFM path extraction and fitting</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMpath.html">SEMpath</a></span><span class="op">(</span><span class="va">pert</span><span class="op">$</span><span class="va">graph</span>, <span class="va">model</span><span class="op">$</span><span class="va">data</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>,</span>
<span>                from <span class="op">=</span> <span class="st">"6647"</span>, <span class="co">#SOD1</span></span>
<span>                to <span class="op">=</span> <span class="st">"4741"</span>, <span class="co">#NEFM</span></span>
<span>                path <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>                verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Path: 6647 -&gt; 4741 size- 11 13 --</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## NLMINB solver ended normally after 1 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 1.612045  srmr: 0.0658653 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node activation: 1.152006e-07 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node inhibition: 0.03466489</span></span></code></pre>
<p><img src="figure-html/unnamed-chunk-5-3.png" style="width:95.0%"></p>
<p>Other possible strategies:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Direct strategy.</span></span>
<span><span class="co"># - Knowledge-based estimation (gnet should be a directed reference network).</span></span>
<span><span class="co"># - Only direct interactions are inferred (d is fixed to 1).</span></span>
<span><span class="co"># - New interactions are inferred from data, during the execution.</span></span>
<span><span class="co"># - A smaller starting beta value is suggested (beta = 0.05)</span></span>
<span></span>
<span><span class="va">model1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelSearch.html">modelSearch</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>,</span>
<span>                      data <span class="op">=</span> <span class="va">data.npn</span>,</span>
<span>                      gnet <span class="op">=</span> <span class="va">kegg</span>, d <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      search <span class="op">=</span> <span class="st">"direct"</span>,</span>
<span>                      beta <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                      method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                      alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                      verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Step1: BAP deconfounding...</span></span>
<span><span class="co">## Step2: DAG estimation...</span></span>
<span><span class="co">## Step3: DAG resize (remove edges/add nodes)...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Done.</span></span></code></pre>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inner strategy.</span></span>
<span><span class="co"># - The reference network is used to validate new interactions and mediators.</span></span>
<span><span class="co"># - Inferred mediators must already belong to the input graph.</span></span>
<span><span class="co"># - Larger d values increase model complexity (suggested: d = 2).</span></span>
<span></span>
<span><span class="va">model2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelSearch.html">modelSearch</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>,</span>
<span>                      data <span class="op">=</span> <span class="va">data.npn</span>,</span>
<span>                      gnet <span class="op">=</span> <span class="va">kegg</span>, d <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                      search <span class="op">=</span> <span class="st">"inner"</span>,</span>
<span>                      beta <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                      method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                      alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                      verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Step1: BAP deconfounding...</span></span>
<span><span class="co">## Step2: DAG estimation...</span></span>
<span><span class="co">## Step3: DAG resize (remove edges/add nodes)...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Done.</span></span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Outer strategy.</span></span>
<span><span class="co"># - Knowledge-based estimation (gnet should be a directed reference network).</span></span>
<span><span class="co"># - Up to (d - 1) mediators can be imported from the reference network.</span></span>
<span><span class="co"># - Larger d values increase model complexity (suggested: d = 2).</span></span>
<span></span>
<span><span class="va">model3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelSearch.html">modelSearch</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>,</span>
<span>                      data <span class="op">=</span> <span class="va">data.npn</span>,</span>
<span>                      gnet <span class="op">=</span> <span class="va">kegg</span>, d <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                      search <span class="op">=</span> <span class="st">"outer"</span>,</span>
<span>                      beta <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                      method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                      alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                      verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Step1: BAP deconfounding...</span></span>
<span><span class="co">## Step2: DAG estimation...</span></span>
<span><span class="co">## Step3: DAG resize (remove edges/add nodes)...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Done.</span></span></code></pre>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Outer strategy adds new nodes from kegg without node labels: </span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">model3</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">model3</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>,</span>
<span>                                <span class="st">'SYMBOL'</span>, <span class="st">'ENTREZID'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># graph models visualization:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">graph</span>, main<span class="op">=</span><span class="st">"basic"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">model1</span><span class="op">$</span><span class="va">graph</span>, main<span class="op">=</span><span class="st">"direct"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">model2</span><span class="op">$</span><span class="va">graph</span>, main<span class="op">=</span><span class="st">"inner"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">model3</span><span class="op">$</span><span class="va">graph</span>, main<span class="op">=</span><span class="st">"outer"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-6-1.png" style="width:95.0%"></p>
<p> </p>
<p>Step 1-3 of modelSearch() with search=“outer” run the following
functions:</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Step 1: Bow-free Acyclic Path (BAP) search.</span></span>
<span></span>
<span><span class="va">BAP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMbap.html">SEMbap</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>,</span>
<span>              data <span class="op">=</span> <span class="va">data.npn</span>,</span>
<span>              method <span class="op">=</span> <span class="st">"BH"</span>, </span>
<span>              alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>              verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Bow-free covariances search. Use method: cggm ...</span></span>
<span><span class="co">## Number of bow-free covariances / df : 220 / 420 </span></span>
<span><span class="co">## Max parent set(S) / Sparsity idx(s) : 10 / 4 </span></span>
<span><span class="co">## Number of clusters / number of nodes: 2 / 31</span></span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Step 2: Directed Acyclic Graph (DAG) estimation.</span></span>
<span></span>
<span><span class="va">DAG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMdag.html">SEMdag</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>,</span>
<span>              data <span class="op">=</span> <span class="va">BAP</span><span class="op">$</span><span class="va">data</span>,</span>
<span>              LO <span class="op">=</span> <span class="st">"TO"</span>,</span>
<span>              beta <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>              lambdas <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>              penalty <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Node Linear Ordering with TO setting</span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Step 3: Graph re-size with external interactome.</span></span>
<span></span>
<span><span class="va">rsg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resizeGraph.html">resizeGraph</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">graph</span>, <span class="va">DAG</span><span class="op">$</span><span class="va">dag.new</span><span class="op">)</span>,</span>
<span>                   gnet <span class="op">=</span> <span class="va">kegg</span>,</span>
<span>                   d <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   v <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  edge set= 1 of 17 edge set= 2 of 17 edge set= 3 of 17 edge set= 4 of 17 edge set= 5 of 17 edge set= 6 of 17 edge set= 7 of 17 edge set= 8 of 17 edge set= 9 of 17 edge set= 10 of 17 edge set= 11 of 17 edge set= 12 of 17 edge set= 13 of 17 edge set= 14 of 17 edge set= 15 of 17 edge set= 16 of 17 edge set= 17 of 17</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  n. edges to be evaluated: 17 </span></span>
<span><span class="co">##  n. edges selected from interactome: 5</span></span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">rsg</span><span class="op">)</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">rsg</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">'SYMBOL'</span>, <span class="st">'ENTREZID'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">rsg</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-7-1.png" style="width:95.0%"></p>
</div>
<div class="section level3">
<h3 id="communities-and-factor-scores-">2.5. Communities and factor scores.<a class="anchor" aria-label="anchor" href="#communities-and-factor-scores-"></a>
</h3>
<p>The modular structure of biological networks could often reveal local
effects and perturbed routes and communities hidden within a larger and
more complex context. <strong>SEMgraph</strong> allows to detect and
estimate these local properties as follows.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Improved ALS model (model$graph) clustering and scoring, using a latent variable</span></span>
<span><span class="co"># "hidden" model (LV), edge betweeness clustering (EBC) algorithm, and a minimum </span></span>
<span><span class="co"># cluster size of 5 nodes.</span></span>
<span><span class="co"># Other clustering algorithms can be exploited (e.g., the walktrap community </span></span>
<span><span class="co"># detection algorithm, WTC) to improve the interpretation of results.</span></span>
<span></span>
<span><span class="va">LV</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterScore.html">clusterScore</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">graph</span>, <span class="va">model</span><span class="op">$</span><span class="va">data</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>,</span>
<span>                   type <span class="op">=</span> <span class="st">"ebc"</span>,</span>
<span>                   HM <span class="op">=</span> <span class="st">"LV"</span>,</span>
<span>                   size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## modularity = 0.4816345 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Community sizes</span></span>
<span><span class="co">##  2  3  1 </span></span>
<span><span class="co">##  6  9 10 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Model converged: TRUE </span></span>
<span><span class="co">## SRMR: 2.629474e-09</span></span></code></pre>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">LV</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  1  2  3 </span></span>
<span><span class="co">## 10  6  9</span></span></code></pre>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">LV</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   lhs op   rhs    est    se      z pvalue ci.lower ci.upper</span></span>
<span><span class="co">## 1 LV1  ~ group  0.043 0.249  0.171  0.864   -0.446    0.532</span></span>
<span><span class="co">## 2 LV2  ~ group -0.471 0.238 -1.975  0.048   -0.938   -0.004</span></span>
<span><span class="co">## 3 LV3  ~ group  0.744 0.287  2.588  0.010    0.181    1.307</span></span>
<span><span class="co">## 4 LV1 ~~   LV1  1.135 0.127  8.944  0.000    0.887    1.384</span></span>
<span><span class="co">## 5 LV2 ~~   LV2  1.037 0.116  8.944  0.000    0.809    1.264</span></span>
<span><span class="co">## 6 LV3 ~~   LV3  1.506 0.168  8.944  0.000    1.176    1.836</span></span></code></pre>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clustering only (no scores calculation)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterGraph.html">clusterGraph</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">graph</span>, type <span class="op">=</span> <span class="st">"ebc"</span>, HM <span class="op">=</span> <span class="st">"LV"</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## modularity = 0.4816345 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Community sizes</span></span>
<span><span class="co">##  2  3  1 </span></span>
<span><span class="co">##  6  9 10</span></span></code></pre>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster plot utility</span></span>
<span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cplot.html">cplot</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">model</span><span class="op">$</span><span class="va">graph</span>, membership <span class="op">=</span> <span class="va">LV</span><span class="op">$</span><span class="va">membership</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-8-1.png" style="width:95.0%"><img src="figure-html/unnamed-chunk-8-2.png" style="width:95.0%"><img src="figure-html/unnamed-chunk-8-3.png" style="width:95.0%"></p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">cg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [[1]]$graph</span></span>
<span><span class="co">## IGRAPH 87f1ee5 DN-- 25 33 -- </span></span>
<span><span class="co">## + attr: name (v/c), color (v/n), label (v/c), M (v/n), color (e/c)</span></span>
<span><span class="co">## + edges from 87f1ee5 (vertex names):</span></span>
<span><span class="co">##  [1] 1432 -&gt;4744  1432 -&gt;4741  5600 -&gt;4744  5600 -&gt;4741  6300 -&gt;4744 </span></span>
<span><span class="co">##  [6] 6300 -&gt;4741  5532 -&gt;4744  5603 -&gt;4741  5603 -&gt;4747  6647 -&gt;84134</span></span>
<span><span class="co">## [11] 6647 -&gt;596   6647 -&gt;4747  6647 -&gt;5535  6647 -&gt;5533  6647 -&gt;79139</span></span>
<span><span class="co">## [16] 54205-&gt;5603  54205-&gt;842   5606 -&gt;1432  5606 -&gt;5603  5608 -&gt;1432 </span></span>
<span><span class="co">## [21] 5608 -&gt;5600  10452-&gt;1432  84134-&gt;5608  4217 -&gt;5608  572  -&gt;54205</span></span>
<span><span class="co">## [26] 572  -&gt;4217  7133 -&gt;4217  7133 -&gt;1616  1616 -&gt;4217  581  -&gt;54205</span></span>
<span><span class="co">## [31] 596  -&gt;54205 596  -&gt;1616  598  -&gt;54205</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$HM1</span></span>
<span><span class="co">## IGRAPH 89acadd DN-- 10 13 -- </span></span>
<span><span class="co">## + attr: name (v/c), color (v/n), label (v/c), M (v/n), color (e/c)</span></span>
<span><span class="co">## + edges from 89acadd (vertex names):</span></span>
<span><span class="co">##  [1] 1432 -&gt;4744 1432 -&gt;4741 5600 -&gt;4744 5600 -&gt;4741 6300 -&gt;4744 6300 -&gt;4741</span></span>
<span><span class="co">##  [7] 5532 -&gt;4744 5603 -&gt;4741 5606 -&gt;1432 5606 -&gt;5603 5608 -&gt;1432 5608 -&gt;5600</span></span>
<span><span class="co">## [13] 10452-&gt;1432</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$HM2</span></span>
<span><span class="co">## IGRAPH 89acb77 DN-- 6 5 -- </span></span>
<span><span class="co">## + attr: name (v/c), color (v/n), label (v/c), M (v/n), color (e/c)</span></span>
<span><span class="co">## + edges from 89acb77 (vertex names):</span></span>
<span><span class="co">## [1] 6647-&gt;84134 6647-&gt;4747  6647-&gt;5535  6647-&gt;5533  6647-&gt;79139</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$HM3</span></span>
<span><span class="co">## IGRAPH 89acbe4 DN-- 9 10 -- </span></span>
<span><span class="co">## + attr: name (v/c), color (v/n), label (v/c), M (v/n), color (e/c)</span></span>
<span><span class="co">## + edges from 89acbe4 (vertex names):</span></span>
<span><span class="co">##  [1] 54205-&gt;842   572  -&gt;54205 572  -&gt;4217  7133 -&gt;4217  7133 -&gt;1616 </span></span>
<span><span class="co">##  [6] 1616 -&gt;4217  581  -&gt;54205 596  -&gt;54205 596  -&gt;1616  598  -&gt;54205</span></span></code></pre>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Colored cluster plot: </span></span>
<span></span>
<span><span class="co">#Set label and node colors</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>,</span>
<span>                            column <span class="op">=</span> <span class="st">'SYMBOL'</span>,</span>
<span>                            keytype <span class="op">=</span> <span class="st">'ENTREZID'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">color</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"lightsalmon"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">color</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">==</span> <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"lightgreen"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">color</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">==</span> <span class="fl">4</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"lightyellow"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">cg</span><span class="op">$</span><span class="va">graph</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-8-4.png" style="width:95.0%"></p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster extraction, fitting, and perturbation evaluation</span></span>
<span></span>
<span><span class="va">cls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractClusters.html">extractClusters</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">model</span><span class="op">$</span><span class="va">graph</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">model</span><span class="op">$</span><span class="va">data</span>,</span>
<span>                       group <span class="op">=</span> <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>,</span>
<span>                       membership <span class="op">=</span> <span class="va">LV</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  cluster= 1 of 3 cluster= 2 of 3 cluster= 3 of 3</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Found 3 clusters with &gt; 5 nodes</span></span></code></pre>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cls</span><span class="op">$</span><span class="va">dfc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   cluster n.nodes n.edges dev_df  srmr V.pv.act V.pv.inh</span></span>
<span><span class="co">## 1     HM1      10      13  2.283 0.068 0.000000 0.014398</span></span>
<span><span class="co">## 2     HM2       6       5  1.895 0.056 0.926485 0.052001</span></span>
<span><span class="co">## 3     HM3       9      10  1.917 0.077 0.122245 0.126563</span></span></code></pre>
<p> </p>
</div>
</div>
<div class="section level2">
<h2 id="frontotemporal-dementia-ftd-data-analysis-">3. Frontotemporal Dementia (FTD) data analysis.<a class="anchor" aria-label="anchor" href="#frontotemporal-dementia-ftd-data-analysis-"></a>
</h2>
<p>The FTD dataset coming with <strong>SEMdata</strong> is a data matrix
of 256 rows (subjects; 150 FTD patients and 150 healthy controls) and
16560 columns (genes) containing the value of the first principal
component of DNAme levels, obtained applying a principal component
analysis to methylated CpG sites within the promoter region, for each
gene (genes with unmethylated CpGs in both conditions were discarded).
This dataset was derived from the study by Li Y. <em>et al.</em>, 2014
(GEO accession: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE53740" class="external-link">GSE53740</a>).</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DNAme PC1 data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ftdDNAme</span><span class="op">$</span><span class="va">pc</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  5934      64778       7419         87       6422</span></span>
<span><span class="co">## GSM1299651 -1.7161298 -2.6574432  1.6903166  1.8872660 -2.7313434</span></span>
<span><span class="co">## GSM1299653 -1.9041925 -1.7208698  3.1574408  1.2142120 -1.2901646</span></span>
<span><span class="co">## GSM1299655 -1.4719359  0.5008334  1.5854376  0.2190518 -0.2458186</span></span>
<span><span class="co">## GSM1299658  0.5574782  1.6449859  0.8602589 -0.8407574  0.8762756</span></span>
<span><span class="co">## GSM1299659 -0.6374026 -0.6834186  0.4256889 -0.3274641 -0.3908066</span></span>
<span><span class="co">## GSM1299660 -0.3370616 -0.1508169 -0.2946679 -1.1251541  0.3142813</span></span></code></pre>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ftdDNAme</span><span class="op">$</span><span class="va">pc1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]   256 16560</span></span></code></pre>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Defining groups</span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">ftdDNAme</span><span class="op">$</span><span class="va">group</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## group</span></span>
<span><span class="co">##   0   1 </span></span>
<span><span class="co">## 151 105</span></span></code></pre>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Nonparanormal transform</span></span>
<span><span class="va">pc1.npn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transformData.html">transformData</a></span><span class="op">(</span><span class="va">ftdDNAme</span><span class="op">$</span><span class="va">pc1</span><span class="op">)</span><span class="op">$</span><span class="va">data</span></span></code></pre></div>
<pre><code><span><span class="co">## Conducting the nonparanormal transformation via shrunkun ECDF...done.</span></span></code></pre>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pc1.npn</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  5934      64778       7419          87       6422</span></span>
<span><span class="co">## GSM1299651 -1.2321143 -1.1917615  1.2965578  1.36667575 -1.6670305</span></span>
<span><span class="co">## GSM1299653 -1.3666758 -0.9197906  2.4600162  1.08068480 -0.9197906</span></span>
<span><span class="co">## GSM1299655 -0.9971162  0.1242948  1.2116994  0.48939828 -0.2555973</span></span>
<span><span class="co">## GSM1299658  0.4893983  0.7142177  0.6766221 -0.42356794  0.5922594</span></span>
<span><span class="co">## GSM1299659 -0.3913234 -0.4127740  0.2047113  0.01487886 -0.4020264</span></span>
<span><span class="co">## GSM1299660 -0.2249971 -0.1342943 -0.2967658 -0.67662208  0.1543331</span></span></code></pre>
<div class="section level3">
<h3 id="gene-set-analysis-gsa-">3.1. Gene Set Analysis (GSA).<a class="anchor" aria-label="anchor" href="#gene-set-analysis-gsa-"></a>
</h3>
<p>In absence of a conceptual model (i.e., one built from an expert’s
indication), the input graph can be inferred from data, literature or
both of them. In the following example, we will take advantage of our
FTD dataset and known FTD-associated pathways from the KEGG database. To
this end, GSA can be used to assess the actual perturbation of known
pathways, given data, and extract those genes (i.e., seeds) underlying
perturbed routes. Seeds can also be defined using knowledge, such as
importing a list of known disease-associated genes, or from other data
sources (e.g. mutational, transcriptional, or epigenetic data).</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load KEGG pathways</span></span>
<span><span class="va">kegg.pathways</span> <span class="op">&lt;-</span> <span class="fu">SEMgraph</span><span class="fu">::</span><span class="va"><a href="../reference/kegg.pathways.html">kegg.pathways</a></span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">kegg.pathways</span><span class="op">)</span> <span class="co">#225</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 225</span></span></code></pre>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kegg.pathways</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "EGFR tyrosine kinase inhibitor resistance"</span></span>
<span><span class="co">## [2] "Endocrine resistance"                     </span></span>
<span><span class="co">## [3] "Antifolate resistance"                    </span></span>
<span><span class="co">## [4] "Platinum drug resistance"                 </span></span>
<span><span class="co">## [5] "mRNA surveillance pathway"                </span></span>
<span><span class="co">## [6] "RNA degradation"</span></span></code></pre>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Known FTD-related pathway selection</span></span>
<span></span>
<span><span class="va">ftd.pathways</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MAPK signaling pathway"</span>,</span>
<span>                  <span class="st">"Protein processing in endoplasmic reticulum"</span>,</span>
<span>                  <span class="st">"Endocytosis"</span>,</span>
<span>                  <span class="st">"Wnt signaling pathway"</span>,</span>
<span>                  <span class="st">"Notch signaling pathway"</span>,</span>
<span>                  <span class="st">"Neurotrophin signaling pathway"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Pathway extraction</span></span>
<span></span>
<span><span class="va">j</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kegg.pathways</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">ftd.pathways</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Gene set analysis (GSA) with 5000 permutations</span></span>
<span></span>
<span><span class="va">ftd.gsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMgsa.html">SEMgsa</a></span><span class="op">(</span><span class="va">kegg.pathways</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>, <span class="va">pc1.npn</span>, <span class="va">group</span>, n_rep <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## k = 1 MAPK signaling pathway </span></span>
<span><span class="co">## k = 2 Protein processing in endoplasmic reticulum </span></span>
<span><span class="co">## k = 3 Endocytosis </span></span>
<span><span class="co">## k = 4 Wnt signaling pathway </span></span>
<span><span class="co">## k = 5 Notch signaling pathway </span></span>
<span><span class="co">## k = 6 Neurotrophin signaling pathway</span></span></code></pre>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ftd.gsa</span><span class="op">$</span><span class="va">gsa</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                             No.nodes No.DEGs     pert</span></span>
<span><span class="co">## Protein processing in endoplasmic reticulum      171      18   up act</span></span>
<span><span class="co">## Endocytosis                                      252      21     &lt;NA&gt;</span></span>
<span><span class="co">## Neurotrophin signaling pathway                   119       3   up act</span></span>
<span><span class="co">## Wnt signaling pathway                            166       0     &lt;NA&gt;</span></span>
<span><span class="co">## MAPK signaling pathway                           294      16   up act</span></span>
<span><span class="co">## Notch signaling pathway                           59       5 down act</span></span>
<span><span class="co">##                                                      pNa          pNi</span></span>
<span><span class="co">## Protein processing in endoplasmic reticulum 2.206778e-09 3.991478e-01</span></span>
<span><span class="co">## Endocytosis                                 1.468533e-08 1.193403e-05</span></span>
<span><span class="co">## Neurotrophin signaling pathway              4.480197e-06 2.574927e-02</span></span>
<span><span class="co">## Wnt signaling pathway                       5.376131e-06 3.229422e-02</span></span>
<span><span class="co">## MAPK signaling pathway                      1.634371e-05 2.145900e-05</span></span>
<span><span class="co">## Notch signaling pathway                     2.056711e-03 1.666035e-04</span></span>
<span><span class="co">##                                                     PVAL         ADJP</span></span>
<span><span class="co">## Protein processing in endoplasmic reticulum 4.413556e-09 2.648134e-08</span></span>
<span><span class="co">## Endocytosis                                 2.937066e-08 1.762239e-07</span></span>
<span><span class="co">## Neurotrophin signaling pathway              8.960395e-06 5.376237e-05</span></span>
<span><span class="co">## Wnt signaling pathway                       1.075226e-05 6.451357e-05</span></span>
<span><span class="co">## MAPK signaling pathway                      3.268743e-05 1.961246e-04</span></span>
<span><span class="co">## Notch signaling pathway                     3.332071e-04 1.999242e-03</span></span></code></pre>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Seed extraction</span></span>
<span></span>
<span><span class="va">seed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">ftd.gsa</span><span class="op">$</span><span class="va">DEG</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">seed</span><span class="op">)</span> <span class="co">#60</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 60</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="network-weighting-and-perturbed-backbone-extraction-">3.2. Network weighting and perturbed backbone extraction.<a class="anchor" aria-label="anchor" href="#network-weighting-and-perturbed-backbone-extraction-"></a>
</h3>
<p>While seeds pinpoint nodes with specific properties, edges can be
weighted on the base of quantitative and phenotype data to define
preferential ways of information exchange (e.g., perturbation
propagation) through the network. <strong>SEMgraph</strong> offers
different alternatives to generate data-driven weights, but the fastest
of them is based on the Fisher’s “r-to-z” method, to test the group
difference between correlation coefficients of pairs of interacting
nodes (Fisher, 1915). Both seeds and weights can be used to extract the
perturbed core(s) of a network to highlight its disease-associated
components.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># KEGG interactome weighting</span></span>
<span></span>
<span><span class="co">#load KEGG interactome</span></span>
<span><span class="va">kegg</span><span class="op">&lt;-</span> <span class="fu">SEMgraph</span><span class="fu">::</span><span class="va"><a href="../reference/kegg.html">kegg</a></span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">kegg</span><span class="op">)</span> <span class="co">#DNW- 5007 44755</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH b46d4ee DNW- 5007 44755 -- </span></span>
<span><span class="co">## + attr: name (v/c), weight (e/n)</span></span></code></pre>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">W</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/weightGraph.html">weightGraph</a></span><span class="op">(</span><span class="va">kegg</span>, <span class="va">pc1.npn</span>, <span class="va">group</span>, method <span class="op">=</span> <span class="st">"r2z"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">W</span><span class="op">)</span> <span class="co">#DNW- 4242 34975 </span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH 9aa7cef DNW- 4242 34975 -- </span></span>
<span><span class="co">## + attr: name (v/c), pv (v/n), zsign (v/n), weight (e/n), zsign (e/n),</span></span>
<span><span class="co">## | pv (e/n)</span></span></code></pre>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perturbed backbone as a Steiner tree (Kou's algorithm)</span></span>
<span><span class="co"># A Steiner tree is the minimum cost (distance) graph</span></span>
<span><span class="co"># including all the specified seeds.</span></span>
<span></span>
<span><span class="va">ST</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMtree.html">SEMtree</a></span><span class="op">(</span><span class="va">W</span>, data <span class="op">=</span> <span class="va">pc1.npn</span>, seed <span class="op">=</span> <span class="va">seed</span>, type <span class="op">=</span> <span class="st">"ST"</span>, eweight <span class="op">=</span> <span class="st">"pvalue"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ST</span><span class="op">)</span> <span class="co">#DNW- 92 (90 directed + 1 bidirected)</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH 9bf5bc9 DNW- 92 92 -- </span></span>
<span><span class="co">## + attr: pv (v/n), zsign (v/n), name (v/c), color (v/c), weight (e/n),</span></span>
<span><span class="co">## | color (e/c), width (e/n)</span></span></code></pre>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Colored plot: green, seed nodes; white, connection nodes</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">ST</span>, l<span class="op">=</span><span class="st">"fdp"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-11-1.png" style="width:95.0%"></p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perturbation evaluation with raw data</span></span>
<span></span>
<span><span class="va">sem1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span><span class="va">ST</span>, <span class="va">pc1.npn</span>, <span class="va">group</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NLMINB solver ended normally after 15 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 5.894448  srmr: 0.4100293 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node activation: 0.000430902 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node inhibition: 3.476108e-13</span></span></code></pre>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Data deconfounding and perturbation evaluation</span></span>
<span></span>
<span><span class="co"># When a DAG is used for causal network inference, missing edges </span></span>
<span><span class="co"># are often masked by unmeasured confounding variables.</span></span>
<span><span class="co"># This step might reduce this effect through Bow-free Acyclic Path (BAP) </span></span>
<span><span class="co"># search and data deconfounding.</span></span>
<span></span>
<span><span class="va">adj.pc1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMbap.html">SEMbap</a></span><span class="op">(</span><span class="va">ST</span>, <span class="va">pc1.npn</span>, method <span class="op">=</span> <span class="st">"bonferroni"</span>, alpha <span class="op">=</span> <span class="fl">5E-06</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DAG conversion : TRUE</span></span>
<span><span class="co">## Bow-free covariances search. Use method: cggm ...</span></span>
<span><span class="co">## Number of bow-free covariances / df : 1208 / 4096 </span></span>
<span><span class="co">## Max parent set(S) / Sparsity idx(s) : 9 / 4 </span></span>
<span><span class="co">## Number of clusters / number of nodes: 2 / 85</span></span></code></pre>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adj.sem1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span><span class="va">ST</span>, <span class="va">adj.pc1</span><span class="op">$</span><span class="va">data</span>, <span class="va">group</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NLMINB solver ended normally after 8 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 1.614218  srmr: 0.0736439 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node activation: 0.03622072 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node inhibition: 0.01004111</span></span></code></pre>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Colored plot: red/pink: activation, blue/lightblue: repression</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">adj.sem1</span><span class="op">$</span><span class="va">graph</span>, l<span class="op">=</span><span class="st">"fdp"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-11-2.png" style="width:95.0%"></p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Tree agglomerative hierarchical clustering (TAHC).</span></span>
<span><span class="co"># This allow us to detect possible communities within our Steiner tree.</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterGraph.html">clusterGraph</a></span><span class="op">(</span><span class="va">ST</span>, type <span class="op">=</span> <span class="st">"tahc"</span><span class="op">)</span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cplot.html">cplot</a></span><span class="op">(</span><span class="va">ST</span>, membership<span class="op">=</span><span class="va">C</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">cg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       Length Class  Mode</span></span>
<span><span class="co">## graph 92     igraph list</span></span>
<span><span class="co">## HM1   25     igraph list</span></span>
<span><span class="co">## HM2   32     igraph list</span></span>
<span><span class="co">## HM3   22     igraph list</span></span>
<span><span class="co">## HM4   13     igraph list</span></span></code></pre>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Colored plot: set 2nd cluster (HM2), node color and size</span></span>
<span><span class="va">cg2</span> <span class="op">&lt;-</span> <span class="va">cg</span><span class="op">$</span><span class="va">HM2</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg2</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg2</span><span class="op">)</span><span class="op">$</span><span class="va">name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">seed</span>, <span class="st">"green"</span>, <span class="st">"white"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg2</span><span class="op">)</span><span class="op">$</span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fl">3</span><span class="op">*</span><span class="fu"><a href="https://r.igraph.org/reference/degree.html" class="external-link">degree</a></span><span class="op">(</span><span class="va">cg2</span>, mode<span class="op">=</span><span class="st">"total"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert Entrez identifiers to gene symbols</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg2</span><span class="op">)</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">cg2</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">'SYMBOL'</span>, <span class="st">'ENTREZID'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot cluster (green nodes are seed nodes)</span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">cg2</span>, l <span class="op">=</span> <span class="st">"neato"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-11-3.png" style="width:95.0%"></p>
</div>
<div class="section level3">
<h3 id="locating-differentially-connected-genes-">3.3. Locating differentially connected genes.<a class="anchor" aria-label="anchor" href="#locating-differentially-connected-genes-"></a>
</h3>
<p>The SEMgraph differential connected inference (DCI) module enables
the detection of perturbed nodes and edges for large graphs. This module
is useful when the aim is to find a perturbed backbone of essential
disease-associated nodes nodes.</p>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Input graph as the union of FTD KEGG pathways</span></span>
<span></span>
<span><span class="va">gU</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="fu">igraph</span><span class="fu">::</span><span class="va"><a href="https://r.igraph.org/reference/union.html" class="external-link">union</a></span>, <span class="va">kegg.pathways</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">gU</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/properties.html">properties</a></span><span class="op">(</span><span class="va">gU</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## Frequency distribution of graph components</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   n.nodes n.graphs</span></span>
<span><span class="co">## 1       3        2</span></span>
<span><span class="co">## 2       4        1</span></span>
<span><span class="co">## 3       7        2</span></span>
<span><span class="co">## 4       8        1</span></span>
<span><span class="co">## 5      10        2</span></span>
<span><span class="co">## 6      12        2</span></span>
<span><span class="co">## 7     586        1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Percent of vertices in the giant component: 63.8 %</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   is.simple      is.dag is.directed is.weighted </span></span>
<span><span class="co">##        TRUE       FALSE        TRUE       FALSE </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## which.mutual.FALSE  which.mutual.TRUE </span></span>
<span><span class="co">##               3564                  8</span></span></code></pre>
<div class="sourceCode" id="cb145"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gU</span><span class="op">)</span> <span class="co">#DNW- 586 3572</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH a607d02 DN-- 586 3572 -- </span></span>
<span><span class="co">## + attr: name (v/c)</span></span></code></pre>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DCI with the WTC clustering algorithm</span></span>
<span></span>
<span><span class="va">gD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMdci.html">SEMdci</a></span><span class="op">(</span><span class="va">gU</span>, <span class="va">pc1.npn</span>, <span class="va">group</span>, type <span class="op">=</span> <span class="st">"wtc"</span>, method <span class="op">=</span> <span class="st">"BH"</span>, alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## modularity = 0.6539108 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Community sizes</span></span>
<span><span class="co">##  14  17  20  25  18  22  21  11   8   4  13  23  10  16  12  15   9  19  24  26 </span></span>
<span><span class="co">##   2   2   2   2   3   3   4   5   6   7   7   7   8   8   9  13  15  15  25  26 </span></span>
<span><span class="co">##   6   1   3   2   5   7 </span></span>
<span><span class="co">##  27  47  68  71  86 118 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## fit cluster = 1 </span></span>
<span><span class="co">## fit cluster = 2 </span></span>
<span><span class="co">## fit cluster = 3 </span></span>
<span><span class="co">## fit cluster = 5 </span></span>
<span><span class="co">## fit cluster = 6 </span></span>
<span><span class="co">## fit cluster = 7 </span></span>
<span><span class="co">## fit cluster = 9 </span></span>
<span><span class="co">## fit cluster = 15 </span></span>
<span><span class="co">## fit cluster = 19 </span></span>
<span><span class="co">## fit cluster = 24 </span></span>
<span><span class="co">## fit cluster = 26 </span></span>
<span><span class="co">## Done.</span></span></code></pre>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gD</span><span class="op">)</span> <span class="co">#DN-- 74 61</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH a8a5cbd DN-- 74 61 -- </span></span>
<span><span class="co">## + attr: name (v/c)</span></span></code></pre>
<div class="sourceCode" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perturbation evaluation of the 1st component of gD</span></span>
<span></span>
<span><span class="va">gC1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/properties.html">properties</a></span><span class="op">(</span><span class="va">gD</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## Frequency distribution of graph components</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   n.nodes n.graphs</span></span>
<span><span class="co">## 1       2        7</span></span>
<span><span class="co">## 2       3        5</span></span>
<span><span class="co">## 3       5        1</span></span>
<span><span class="co">## 4      14        1</span></span>
<span><span class="co">## 5      26        1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Percent of vertices in the giant component: 35.1 %</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   is.simple      is.dag is.directed is.weighted </span></span>
<span><span class="co">##        TRUE        TRUE        TRUE       FALSE </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## which.mutual.FALSE </span></span>
<span><span class="co">##                 27</span></span></code></pre>
<div class="sourceCode" id="cb153"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sem1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span><span class="va">gC1</span>, <span class="va">pc1.npn</span>, <span class="va">group</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NLMINB solver ended normally after 6 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 12.68914  srmr: 0.3918051 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node activation: 0.2123603 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of node inhibition: 4.865384e-05</span></span></code></pre>
<div class="sourceCode" id="cb155"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sem2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEMrun.html">SEMrun</a></span><span class="op">(</span><span class="va">sem1</span><span class="op">$</span><span class="va">graph</span>, <span class="va">pc1.npn</span>, <span class="va">group</span>, fit <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NLMINB solver ended normally after 2 iterations </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## deviance/df: 5.099692  srmr: 0.1504874 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of edge activation: 0.01959594 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Brown's combined P-value of edge inhibition: 9.001465e-06</span></span></code></pre>
<div class="sourceCode" id="cb157"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Colored DCI plot</span></span>
<span><span class="va">gC1</span> <span class="op">&lt;-</span> <span class="va">sem2</span><span class="op">$</span><span class="va">graph</span></span>
<span></span>
<span><span class="co"># Convert Entrez identifiers to gene symbols</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">gC1</span><span class="op">)</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">gC1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">"SYMBOL"</span>, <span class="st">"ENTREZID"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb159"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set node size</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">gC1</span><span class="op">)</span><span class="op">$</span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fl">1.5</span><span class="op">*</span><span class="fu"><a href="https://r.igraph.org/reference/degree.html" class="external-link">degree</a></span><span class="op">(</span><span class="va">gC1</span>, mode<span class="op">=</span><span class="st">"total"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/gplot.html">gplot</a></span><span class="op">(</span><span class="va">gC1</span>, l<span class="op">=</span><span class="st">"neato"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure-html/unnamed-chunk-12-1.png" style="width:95.0%"></p>
<p>Note: the global statistics suggest a poor fit, and de-correlation
could be performed; however, a significant perturbation between the two
conditions is observed, as shown in the extracted subgraph by pink (beta
&gt; 0: activation) or light blue (beta &lt; 0: repression) nodes, and
red (delta &gt; 0: activation) or blue (delta &lt; 0: repression) edges,
respectively.</p>
<p> </p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p><a href="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/38/20/10.1093_bioinformatics_btac567/1/btac567_supplementary_data.pdf?Expires=1767731027&amp;Signature=o20YJlVgCpAI7ZHJQQw-rFOtX3kKhcHF7SUYuAIe41hFelCUY3q7D5KHk-EbhLQAyDrTQL-MiwuKiPorExwUeGBLQ1wo5XwDOfrTEiqXxl~CW0lOK4peJjn9fBkI95Hk7cvKDd1OQud9R8pQp32~VykXlIDZ3NnRk-BUCRR4fF8Yj0lmZEAevaTJ27OtMBq4TkpFxKEpy9P3eB3ufmVs6hXsWnfm5Eh4vDSf9yH8O4Wj498Su1BB6V2MvkScsKhbHkSJDAVBaCOoI1q4EDUwr28shYzKGcohnxPlTcK9KYne~bie4YUZSB4anH0SweKv~7gNGyy5B6zVen3WfvOF9A__&amp;Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" class="external-link">Supplementary
material</a> of Grassi M, Palluzzi F, Tarantino B. <strong>SEMgraph: an
R package for causal network inference of high-throughput data with
structural equation models</strong>. Bioinformatics, 2022 Aug 30;
38(20):btac567. <a href="https://doi.org/10.1093/bioinformatics/btac567" class="external-link uri">https://doi.org/10.1093/bioinformatics/btac567</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mario Grassi, Fernando Palluzzi, Barbara Tarantino.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
