<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fit a graph as a Structural Equation Model (SEM) — SEMrun • SEMgraph</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fit a graph as a Structural Equation Model (SEM) — SEMrun" />
<meta property="og:description" content="SEMrun() converts a (directed, undirected, or mixed)
graph to a SEM and fits it. If a binary group variable (i.e., case/control)
is present, node-level or edge-level perturbation is evaluated.
This function can handle loop-containing models, although multiple
links between the same two nodes (including self-loops and mutual
interactions) and bows (i.e., a directed and a bidirected link between
two nodes) are not allowed." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SEMgraph</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/fernandoPalluzzi/SEMgraph/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit a graph as a Structural Equation Model (SEM)</h1>
    <small class="dont-index">Source: <a href='https://github.com/fernandoPalluzzi/SEMgraph/blob/master/R/semPaths.R'><code>R/semPaths.R</code></a></small>
    <div class="hidden name"><code>SEMrun.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>SEMrun()</code> converts a (directed, undirected, or mixed)
graph to a SEM and fits it. If a binary group variable (i.e., case/control)
is present, node-level or edge-level perturbation is evaluated.
This function can handle loop-containing models, although multiple
links between the same two nodes (including self-loops and mutual
interactions) and bows (i.e., a directed and a bidirected link between
two nodes) are not allowed.</p>
    </div>

    <pre class="usage"><span class='fu'>SEMrun</span><span class='op'>(</span>
  <span class='va'>graph</span>,
  <span class='va'>data</span>,
  group <span class='op'>=</span> <span class='cn'>NULL</span>,
  fit <span class='op'>=</span> <span class='fl'>0</span>,
  algo <span class='op'>=</span> <span class='st'>"lavaan"</span>,
  start <span class='op'>=</span> <span class='cn'>NULL</span>,
  limit <span class='op'>=</span> <span class='fl'>100</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>graph</th>
      <td><p>An igraph object.</p></td>
    </tr>
    <tr>
      <th>data</th>
      <td><p>A matrix whith rows corresponding to subjects, and
columns to graph nodes (variables).</p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>A binary vector. This vector must be as long as the
number of subjects. Each vector element must be 1 for cases and 0
for control subjects. If <code>NULL</code> (default), group influence will
not be considered.</p></td>
    </tr>
    <tr>
      <th>fit</th>
      <td><p>A numeric value indicating the SEM fitting mode.
If <code>fit = 0</code> (default), no group effect is considered.
If <code>fit = 1</code>, a "common" model is used to evaluate group effects
on graph nodes.
If <code>fit = 2</code>, a two-group model is used to evaluate group effects
on graph edges.</p></td>
    </tr>
    <tr>
      <th>algo</th>
      <td><p>MLE method used for SEM fitting. If <code>algo = "lavaan"</code>
(default), the SEM will be fitted using the NLMINB solver from
<code>lavaan</code> R package, with standard errors derived from the expected
Fisher information matrix. If <code>algo = "ricf"</code>, the model is fitted
via residual iterative conditional fitting (RICF; Drton et al. 2009).
If <code>algo = "cggm"</code>, model fitting is based on constrained Gaussian
Graphical Modeling (GGM) and de-sparsified glasso estimator
(Williams, 2020).</p></td>
    </tr>
    <tr>
      <th>start</th>
      <td><p>Starting value of SEM parameters for <code>algo = "lavaan"</code>.
If start is <code>NULL</code> (default), the algorithm will determine the
starting values. If start is a numeric value, it will be used as a
scaling factor for the edge weights in the graph object (graph attribute
<code>E(graph)$weight</code>).
For instance, a scaling factor is useful when weights have fixed values
(e.g., 1 for activated, -1 for repressed, and 0 for unchanged interaction).
Fixed values may compromise model fitting, and scaling them is a safe
option to avoid this problem. As a rule of thumb, to our experience,
<code>start = 0.1</code> generally performs well with -1, 0, 1 weights.</p></td>
    </tr>
    <tr>
      <th>limit</th>
      <td><p>An integer value corresponding to the network size
(i.e., number of nodes). Beyond this limit, the execution under
<code>algo = "lavaan"</code> will be ridirected to <code>algo = "ricf"</code>, if
fit is either 0 or 1, or to <code>algo = "ggm"</code>, if <code>fit = 2</code>.
This redirection is necessary to reduce the computational demand of
standard error estimation by lavaan. Increasing this number will
enforce lavaan execution when <code>algo = "lavaan"</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Currently ignored.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of 5 objects:</p><ol>
<li><p>"fit", SEM fitted lavaan, ricf, or ggmncv object,
depending on the MLE method specified by the <code>algo</code> argument;</p></li>
<li><p>"gest" or "dest", a data.frame of node-specific
("gest") or edge-specific ("dest") group effect estimates and P-values;</p></li>
<li><p>"model", SEM model as a string if <code>algo = "lavaan"</code>,
and <code>NULL</code> otherwise;</p></li>
<li><p>"graph", the induced subgraph of the input network mapped
on data variables. Graph edges (i.e., direct effects) with P-value &lt; 0.05
will be highlighted in red (beta &gt; 0) or blue (beta &lt; 0). If a group
vector is given, nodes with significant group effect (P-value &lt; 0.05)
will be red-shaded (beta &gt; 0) or lightblue-shaded (beta &lt; 0);</p></li>
<li><p>"dataXY", input data subset mapping graph nodes, plus
group at the first column (if no group is specified, this column will
take NA values).</p></li>
</ol>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>SEMrun maps data onto the input graph and converts it into a
SEM. Directed connections (X -&gt; Y) are interpreted as direct causal
effects, while undirected, mutual, and bidirected connections are
converted into model covariances. SEMrun output contains different sets
of parameter estimates. Beta coefficients (i.e., direct effects) are
estimated from directed interactions and residual covariances (psi
coefficients) from bidirected, undirected, or mutual interactions.
If a group variable is given, exogenous group effects on nodes (gamma
coefficients) will be estimated. This will also lead to the estimation
of a set of aggregated group effects, if <code>algo = "ricf"</code> (see
<code><a href='SEMgsa.html'>SEMgsa</a></code>).
By default, maximum likelihood parameter estimates and P-values for
parameter sets are computed by conventional z-test (= estimate/SE),
and fits it through the <code><a href='https://rdrr.io/pkg/lavaan/man/lavaan.html'>lavaan</a></code> function, via
Maximum Likelihood Estimation (estimator = "ML", default estimator in
<code><a href='https://rdrr.io/pkg/lavaan/man/lavOptions.html'>lavOptions</a></code>).
In case of high dimensionality (n.variables &gt;&gt; n.subjects), the covariance
matrix could not be semi-definite positive and thus parameter estimates
could not be done. If this happens, covariance matrix regularization
is enabled using the James-Stein-type shrinkage estimator implemented
in the function <code><a href='https://rdrr.io/pkg/corpcor/man/pcor.shrink.html'>pcor.shrink</a></code> of corpcor R package.
Argument <code>fit</code> determines how group influence is evaluated in the
model, as absent (<code>fit = 0</code>), node perturbation (<code>fit = 1</code>),
or edge perturbation (<code>fit = 2</code>). When <code>fit = 1</code>, the group
is modeled as an exogenous variable, influencing all the other graph
nodes. When <code>fit = 2</code>, SEMrun estimates the differences
of the beta and/or psi coefficients (network edges) between groups.
This is equivalent to fit a separate model for cases and controls,
as opposed to one common model perturbed by the exogenous group effect.
Once fitted, the two models are then compared to assess significant
edge (i.e., direct effect) differences (d = beta1 - beta0).
P-values for parameter sets are computed by z-test (= d/SE), through
<code><a href='https://rdrr.io/pkg/lavaan/man/lavaan.html'>lavaan</a></code>. As an alternative to standard P-value
calculation, SEMrun may use either RICF (randomization P-values) or
GGM (de-sparsified P-values) methods. These algorithms are much faster
than <code><a href='https://rdrr.io/pkg/lavaan/man/lavaan.html'>lavaan</a></code> in case of large input graphs.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Pearl J (1998). Graphs, Causality, and Structural Equation Models.
Sociological Methods &amp; Research., 27(2):226-284.
&lt;https://doi.org/10.1177/0049124198027002004&gt;</p>
<p>Yves Rosseel (2012). lavaan: An R Package for Structural Equation
Modeling. Journal of Statistical Software, 48(2): 1-36.
&lt;https://www.jstatsoft.org/v48/i02/&gt;</p>
<p>Pepe D, Grassi M (2014). Investigating perturbed pathway modules
from gene expression data via Structural Equation Models. BMC
Bioinformatics, 15: 132.
&lt;https://doi.org/10.1186/1471-2105-15-132&gt;</p>
<p>Drton M, Eichler M, Richardson TS (2009). Computing Maximum Likelihood
Estimated in Recursive Linear Models with Correlated Errors.
Journal of Machine Learning Research, 10(Oct): 2329-2348.
&lt;https://www.jmlr.org/papers/volume10/drton09a/drton09a.pdf&gt;</p>
<p>Larson JL and Owen AB (2015). Moment based gene set tests. BMC
Bioinformatics, 16: 132. &lt;https://doi.org/10.1186/s12859-015-0571-7&gt;</p>
<p>Palluzzi F, Grassi M (2021). SEMgraph: An R Package for Causal Network
Analysis of High-Throughput Data with Structural Equation Models.
&lt;arXiv:2103.08332&gt;</p>
<p>Williams D (2020). GGMncv: Gaussian Graphical Models with Non-Convex
Penalties. R package version 1.1.0.
&lt;https://CRAN.R-project.org/package=GGMncv&gt;</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>See <code><a href='https://rdrr.io/pkg/ggm/man/fitAncestralGraph.html'>fitAncestralGraph</a></code> for RICF algorithm
details, <code><a href='https://rdrr.io/pkg/flip/man/flip.html'>flip</a></code> for randomization P-values, and
<code><a href='https://rdrr.io/pkg/GGMncv/man/constrained.html'>constrained</a></code> for constrained GGM, and
<code><a href='https://rdrr.io/pkg/GGMncv/man/inference.html'>inference</a></code> for de-sparsified P-values.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Mario Grassi <a href='mailto:mario.grassi@unipv.it'>mario.grassi@unipv.it</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>#### Model fitting (no group effect)</span>

<span class='va'>sem0</span> <span class='op'>&lt;-</span> <span class='fu'>SEMrun</span><span class='op'>(</span>graph <span class='op'>=</span> <span class='va'>sachs</span><span class='op'>$</span><span class='va'>graph</span>, data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log</a></span><span class='op'>(</span><span class='va'>sachs</span><span class='op'>$</span><span class='va'>pkc</span><span class='op'>)</span>, algo <span class='op'>=</span> <span class='st'>"lavaan"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: lavaan WARNING:</span>
#&gt; <span class='warning'>    model syntax contains variance/covariance/intercept formulas</span>
#&gt; <span class='warning'>    involving (an) exogenous variable(s): [zPIP3 zPKA]; These</span>
#&gt; <span class='warning'>    variables will now be treated as random introducing additional</span>
#&gt; <span class='warning'>    free parameters. If you wish to treat those variables as fixed,</span>
#&gt; <span class='warning'>    remove these formulas from the model syntax. Otherwise, consider</span>
#&gt; <span class='warning'>    adding the fixed.x = FALSE option.</span></div><div class='output co'>#&gt; NLMINB solver ended normally after 14 iterations 
#&gt; 
#&gt; deviance/df: 66.93929  srmr: 0.0894731 
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>sem0</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>)</span>
</div><div class='output co'>#&gt; lavaan 0.6-8 ended normally after 14 iterations
#&gt; 
#&gt;   Estimator                                         ML
#&gt;   Optimization method                           NLMINB
#&gt;   Number of model parameters                        30
#&gt;                                                       
#&gt;   Number of observations                          1766
#&gt;                                                       
#&gt; Model Test User Model:
#&gt;                                                       
#&gt;   Test statistic                              2409.815
#&gt;   Degrees of freedom                                36
#&gt;   P-value (Chi-square)                           0.000
#&gt; 
#&gt; Parameter Estimates:
#&gt; 
#&gt;   Standard errors                             Standard
#&gt;   Information                                 Observed
#&gt;   Observed information based on                Hessian
#&gt; 
#&gt; Regressions:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;   zAkt ~                                              
#&gt;     zPIP3            -0.065    0.020   -3.251    0.001
#&gt;     zPKA              0.546    0.020   27.454    0.000
#&gt;   zErk ~                                              
#&gt;     zMek             -0.021    0.021   -0.974    0.330
#&gt;     zPKA              0.456    0.021   21.506    0.000
#&gt;   zJnk ~                                              
#&gt;     zPKA              0.093    0.024    3.940    0.000
#&gt;     zPKC              0.128    0.024    5.439    0.000
#&gt;   zMek ~                                              
#&gt;     zPKA              0.011    0.018    0.615    0.539
#&gt;     zPKC             -0.006    0.017   -0.321    0.748
#&gt;     zRaf              0.683    0.018   38.931    0.000
#&gt;   zP38 ~                                              
#&gt;     zPKA             -0.018    0.018   -0.988    0.323
#&gt;     zPKC              0.646    0.018   35.504    0.000
#&gt;   zPIP2 ~                                             
#&gt;     zPIP3             0.489    0.020   24.698    0.000
#&gt;     zPlcg             0.227    0.020   11.441    0.000
#&gt;   zPKC ~                                              
#&gt;     zPIP2             0.020    0.025    0.782    0.434
#&gt;     zPlcg             0.014    0.025    0.570    0.568
#&gt;   zPlcg ~                                             
#&gt;     zPIP3             0.207    0.023    8.895    0.000
#&gt;   zRaf ~                                              
#&gt;     zPKA             -0.124    0.024   -5.261    0.000
#&gt;     zPKC             -0.037    0.024   -1.587    0.113
#&gt; 
#&gt; Covariances:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;   zPIP3 ~~                                            
#&gt;     zPKA             -0.103    0.024   -4.306    0.000
#&gt; 
#&gt; Variances:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;    .zAkt              0.690    0.023   29.715    0.000
#&gt;    .zErk              0.790    0.027   29.715    0.000
#&gt;    .zJnk              0.973    0.033   29.715    0.000
#&gt;    .zMek              0.534    0.018   29.715    0.000
#&gt;    .zP38              0.583    0.020   29.715    0.000
#&gt;    .zPIP2             0.663    0.022   29.715    0.000
#&gt;    .zPKC              0.999    0.034   29.715    0.000
#&gt;    .zPlcg             0.957    0.032   29.715    0.000
#&gt;    .zRaf              0.982    0.033   29.715    0.000
#&gt;     zPIP3             0.999    0.034   29.715    0.000
#&gt;     zPKA              0.999    0.034   29.715    0.000
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'>parameterEstimates</span><span class='op'>(</span><span class='va'>sem0</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;    lhs op   rhs    est    se      z pvalue ci.lower ci.upper
#&gt; 1 zAkt  ~ zPIP3 -0.065 0.020 -3.251  0.001   -0.104   -0.026
#&gt; 2 zAkt  ~  zPKA  0.546 0.020 27.454  0.000    0.507    0.585
#&gt; 3 zErk  ~  zMek -0.021 0.021 -0.974  0.330   -0.062    0.021
#&gt; 4 zErk  ~  zPKA  0.456 0.021 21.506  0.000    0.415    0.498
#&gt; 5 zJnk  ~  zPKA  0.093 0.024  3.940  0.000    0.047    0.139
#&gt; 6 zJnk  ~  zPKC  0.128 0.024  5.439  0.000    0.082    0.174</div><div class='input'>
<span class='va'>sem0</span> <span class='op'>&lt;-</span> <span class='fu'>SEMrun</span><span class='op'>(</span>graph <span class='op'>=</span> <span class='va'>sachs</span><span class='op'>$</span><span class='va'>graph</span>, data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log</a></span><span class='op'>(</span><span class='va'>sachs</span><span class='op'>$</span><span class='va'>pkc</span><span class='op'>)</span>, algo <span class='op'>=</span> <span class='st'>"ricf"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; RICF solver ended normally after 2 iterations 
#&gt; 
#&gt; deviance/df: 66.93929  srmr: 0.0894731 
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>sem0</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>)</span>
</div><div class='output co'>#&gt; RICF solver ended normally after 2 iterations 
#&gt; 
#&gt;   Estimator                                       ML 
#&gt;   Optimization method                             RICF 
#&gt; 
#&gt;   Number of free parameters                       30 
#&gt; 
#&gt;   Number of observations                          1766 
#&gt; 
#&gt; Model Test User Model
#&gt; 
#&gt;   Test statistic (Deviance)                       2409.815 
#&gt;   Degrees of freedom (df)                         36 
#&gt;   Deviance/df                                     66.939 
#&gt;   Standardized Root Mean Square Residual (srmr)   0.089 
#&gt; 
#&gt; Parameter Estimates:
#&gt; 
#&gt; Regressions: 
#&gt; 
#&gt;     lhs op  rhs    est
#&gt; 1   Akt  ~ PIP3 -0.065
#&gt; 2   Akt  ~  PKA  0.546
#&gt; 3   Erk  ~  Mek -0.021
#&gt; 4   Erk  ~  PKA  0.456
#&gt; 5   Jnk  ~  PKA  0.093
#&gt; 6   Jnk  ~  PKC  0.128
#&gt; 7   Mek  ~  Raf  0.683
#&gt; 8   Mek  ~  PKA  0.011
#&gt; 9   Mek  ~  PKC -0.006
#&gt; 10  P38  ~  PKA -0.018
#&gt; 11  P38  ~  PKC  0.646
#&gt; 12 PIP2  ~ Plcg  0.227
#&gt; 13 PIP2  ~ PIP3  0.489
#&gt; 14  PKC  ~ Plcg  0.014
#&gt; 15  PKC  ~ PIP2  0.020
#&gt; 16 Plcg  ~ PIP3  0.207
#&gt; 17  Raf  ~  PKA -0.124
#&gt; 18  Raf  ~  PKC -0.037
#&gt; 
#&gt; Covariances: 
#&gt; 
#&gt;   lhs op  rhs    est
#&gt; 1 PKA ~~ PIP3 -0.103
#&gt; 
#&gt; Variances: 
#&gt; 
#&gt;     lhs op  rhs   est
#&gt; 1   Akt ~~  Akt 0.691
#&gt; 2   Erk ~~  Erk 0.790
#&gt; 3   Jnk ~~  Jnk 0.974
#&gt; 4   Mek ~~  Mek 0.535
#&gt; 5   P38 ~~  P38 0.583
#&gt; 6  PIP2 ~~ PIP2 0.663
#&gt; 7  PIP3 ~~ PIP3 1.000
#&gt; 8   PKA ~~  PKA 1.000
#&gt; 9   PKC ~~  PKC 0.999
#&gt; 10 Plcg ~~ Plcg 0.957
#&gt; 11  Raf ~~  Raf 0.983
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>sem0</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>$</span><span class='va'>parameterEstimates</span><span class='op'>)</span>
</div><div class='output co'>#&gt; $Reg
#&gt;     lhs op  rhs          est
#&gt; 8   Akt  ~ PIP3 -0.064637571
#&gt; 12  Akt  ~  PKA  0.545827288
#&gt; 2   Erk  ~  Mek -0.020662182
#&gt; 11  Erk  ~  PKA  0.456197198
#&gt; 14  Jnk  ~  PKA  0.092697998
#&gt; 18  Jnk  ~  PKC  0.127962060
#&gt; 1   Mek  ~  Raf  0.683280129
#&gt; 10  Mek  ~  PKA  0.010804652
#&gt; 16  Mek  ~  PKC -0.005594917
#&gt; 13  P38  ~  PKA -0.017995367
#&gt; 17  P38  ~  PKC  0.646469780
#&gt; 3  PIP2  ~ Plcg  0.226646823
#&gt; 7  PIP2  ~ PIP3  0.489264356
#&gt; 4   PKC  ~ Plcg  0.014362915
#&gt; 5   PKC  ~ PIP2  0.019696273
#&gt; 6  Plcg  ~ PIP3  0.207077521
#&gt; 9   Raf  ~  PKA -0.124349458
#&gt; 15  Raf  ~  PKC -0.037496729
#&gt; 
#&gt; $Cov
#&gt;    lhs op  rhs        est
#&gt; 24 PKA ~~ PIP3 -0.1030054
#&gt; 
#&gt; $Var
#&gt;     lhs op  rhs       est
#&gt; 26  Akt ~~  Akt 0.6906263
#&gt; 25  Erk ~~  Erk 0.7900222
#&gt; 30  Jnk ~~  Jnk 0.9735446
#&gt; 20  Mek ~~  Mek 0.5345122
#&gt; 29  P38 ~~  P38 0.5832125
#&gt; 22 PIP2 ~~ PIP2 0.6633259
#&gt; 23 PIP3 ~~ PIP3 1.0000000
#&gt; 27  PKA ~~  PKA 1.0000000
#&gt; 28  PKC ~~  PKC 0.9992202
#&gt; 21 Plcg ~~ Plcg 0.9571189
#&gt; 19  Raf ~~  Raf 0.9825462
#&gt; </div><div class='input'>
<span class='va'>sem0</span> <span class='op'>&lt;-</span> <span class='fu'>SEMrun</span><span class='op'>(</span>graph <span class='op'>=</span> <span class='va'>sachs</span><span class='op'>$</span><span class='va'>graph</span>, data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log</a></span><span class='op'>(</span><span class='va'>sachs</span><span class='op'>$</span><span class='va'>pkc</span><span class='op'>)</span>, algo <span class='op'>=</span> <span class='st'>"cggm"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GGM (constrained) solver ended normally after 0 iterations 
#&gt; 
#&gt; deviance/df: 66.86359  srmr: 0.1840727 
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>sem0</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GGM (constrained) solver ended normally after 0 iterations 
#&gt; 
#&gt;   Estimator                                       ML 
#&gt;   Optimization method                             CGGM 
#&gt; 
#&gt;   Number of free parameters                       30 
#&gt; 
#&gt;   Number of observations                          1766 
#&gt; 
#&gt; Model Test User Model
#&gt; 
#&gt;   Test statistic (Deviance)                       2407.089 
#&gt;   Degrees of freedom (df)                         36 
#&gt;   Deviance/df                                     66.864 
#&gt;   Standardized Root Mean Square Residual (srmr)   0.184 
#&gt; 
#&gt; Parameter Estimates:
#&gt; 
#&gt; Regressions: 
#&gt; 
#&gt;     lhs op  rhs    est pvalue
#&gt; 1   Akt  ~ PIP3 -0.065  0.002
#&gt; 2   Akt  ~  PKA  0.546  0.000
#&gt; 3   Erk  ~  Mek -0.021  0.030
#&gt; 4   Erk  ~  PKA  0.456  0.073
#&gt; 5   Jnk  ~  PKA  0.097  0.633
#&gt; 6   Jnk  ~  PKC  0.131  0.726
#&gt; 7   Mek  ~  Raf  0.683  0.000
#&gt; 8   Mek  ~  PKA  0.017  0.178
#&gt; 9   Mek  ~  PKC -0.005  0.886
#&gt; 10  P38  ~  PKA  0.007  0.863
#&gt; 11  P38  ~  PKC  0.645  0.000
#&gt; 12 PIP2  ~ Plcg  0.226  0.000
#&gt; 13 PIP2  ~ PIP3  0.489  0.000
#&gt; 14  PKC  ~ Plcg  0.008  0.872
#&gt; 15  PKC  ~ PIP2  0.012  0.880
#&gt; 16 Plcg  ~ PIP3  0.044  0.041
#&gt; 17  Raf  ~  PKA -0.075  0.398
#&gt; 18  Raf  ~  PKC -0.019  0.585
#&gt; 
#&gt; Covariances: 
#&gt; 
#&gt;    lhs op rhs    est pvalue
#&gt; 1 PIP3 ~~ PKA -0.103  0.379
#&gt; 
#&gt; Variances: 
#&gt; 
#&gt;     lhs op  rhs   est pvalue
#&gt; 1   Akt ~~  Akt 0.691      0
#&gt; 2   Erk ~~  Erk 0.790      0
#&gt; 3   Jnk ~~  Jnk 0.974      0
#&gt; 4   Mek ~~  Mek 0.535      0
#&gt; 5   P38 ~~  P38 0.584      0
#&gt; 6  PIP2 ~~ PIP2 0.663      0
#&gt; 7  PIP3 ~~ PIP3 1.000      0
#&gt; 8   PKA ~~  PKA 1.000      0
#&gt; 9   PKC ~~  PKC 0.999      0
#&gt; 10 Plcg ~~ Plcg 0.984      0
#&gt; 11  Raf ~~  Raf 0.985      0
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>sem0</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>$</span><span class='va'>parameterEstimates</span><span class='op'>)</span>
</div><div class='output co'>#&gt; $Reg
#&gt;     lhs op  rhs          est        pvalue
#&gt; 8   Akt  ~ PIP3 -0.064917127  1.813857e-03
#&gt; 12  Akt  ~  PKA  0.545580110  1.923273e-09
#&gt; 2   Erk  ~  Mek -0.020537125  3.023840e-02
#&gt; 11  Erk  ~  PKA  0.455766193  7.336207e-02
#&gt; 14  Jnk  ~  PKA  0.097276265  6.326038e-01
#&gt; 18  Jnk  ~  PKC  0.131322957  7.262377e-01
#&gt; 1   Mek  ~  Raf  0.683057189 1.006115e-122
#&gt; 10  Mek  ~  PKA  0.017103153  1.780280e-01
#&gt; 16  Mek  ~  PKC -0.005344735  8.855709e-01
#&gt; 13  P38  ~  PKA  0.007001167  8.628011e-01
#&gt; 17  P38  ~  PKC  0.645274212 3.634077e-109
#&gt; 3  PIP2  ~ Plcg  0.226127321  1.554591e-26
#&gt; 7  PIP2  ~ PIP3  0.489389920  8.422860e-81
#&gt; 4   PKC  ~ Plcg  0.008073818  8.717163e-01
#&gt; 5   PKC  ~ PIP2  0.012110727  8.797994e-01
#&gt; 6  Plcg  ~ PIP3  0.043672014  4.113091e-02
#&gt; 9   Raf  ~  PKA -0.074524313  3.979705e-01
#&gt; 15  Raf  ~  PKC -0.019027484  5.848711e-01
#&gt; 
#&gt; $Cov
#&gt;     lhs op rhs    est    pvalue
#&gt; 26 PIP3 ~~ PKA -0.103 0.3792156
#&gt; 
#&gt; $Var
#&gt;     lhs op  rhs       est pvalue
#&gt; 25  Akt ~~  Akt 0.6906264  1e-09
#&gt; 24  Erk ~~  Erk 0.7900224  1e-09
#&gt; 30  Jnk ~~  Jnk 0.9735788  1e-09
#&gt; 20  Mek ~~  Mek 0.5345526  1e-09
#&gt; 29  P38 ~~  P38 0.5838350  1e-09
#&gt; 22 PIP2 ~~ PIP2 0.6633261  1e-09
#&gt; 23 PIP3 ~~ PIP3 1.0000000  1e-09
#&gt; 27  PKA ~~  PKA 1.0000000  1e-09
#&gt; 28  PKC ~~  PKC 0.9993486  1e-09
#&gt; 21 Plcg ~~ Plcg 0.9838203  1e-09
#&gt; 19  Raf ~~  Raf 0.9854853  1e-09
#&gt; </div><div class='input'>
<span class='co'># Graphs</span>
<span class='fu'><a href='gplot.html'>gplot</a></span><span class='op'>(</span><span class='va'>sem0</span><span class='op'>$</span><span class='va'>graph</span>, main <span class='op'>=</span> <span class='st'>"edge differences"</span><span class='op'>)</span>
</div><div class='img'><img src='SEMrun-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>sem0</span><span class='op'>$</span><span class='va'>graph</span>, layout <span class='op'>=</span> <span class='va'>layout.circle</span>, main <span class='op'>=</span> <span class='st'>"edge differences"</span><span class='op'>)</span>
</div><div class='img'><img src='SEMrun-2.png' alt='' width='700' height='433' /></div><div class='input'>

<span class='co'>#### Model fitting (common model, group effect on nodes)</span>

<span class='va'>sem1</span> <span class='op'>&lt;-</span> <span class='fu'>SEMrun</span><span class='op'>(</span>graph <span class='op'>=</span> <span class='va'>sachs</span><span class='op'>$</span><span class='va'>graph</span>, data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log</a></span><span class='op'>(</span><span class='va'>sachs</span><span class='op'>$</span><span class='va'>pkc</span><span class='op'>)</span>,
               group <span class='op'>=</span> <span class='va'>sachs</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span>
</div><div class='output co'>#&gt; NLMINB solver ended normally after 18 iterations 
#&gt; 
#&gt; deviance/df: 61.84434  srmr: 0.07009673 
#&gt; 
#&gt; Brown's combined P-value of node activation: 0 
#&gt; 
#&gt; Brown's combined P-value of node inhibition: 0 
#&gt; </div><div class='input'>
<span class='co'># Fitting summaries</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>sem1</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>)</span>
</div><div class='output co'>#&gt; lavaan 0.6-8 ended normally after 18 iterations
#&gt; 
#&gt;   Estimator                                         ML
#&gt;   Optimization method                           NLMINB
#&gt;   Number of model parameters                        41
#&gt;                                                       
#&gt;   Number of observations                          1766
#&gt;                                                       
#&gt; Model Test User Model:
#&gt;                                                       
#&gt;   Test statistic                              2226.396
#&gt;   Degrees of freedom                                36
#&gt;   P-value (Chi-square)                           0.000
#&gt; 
#&gt; Parameter Estimates:
#&gt; 
#&gt;   Standard errors                             Standard
#&gt;   Information                                 Observed
#&gt;   Observed information based on                Hessian
#&gt; 
#&gt; Regressions:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;   zAkt ~                                              
#&gt;     group             0.150    0.021    7.057    0.000
#&gt;   zErk ~                                              
#&gt;     group             0.179    0.022    8.046    0.000
#&gt;   zJnk ~                                              
#&gt;     group             0.333    0.023   14.279    0.000
#&gt;   zMek ~                                              
#&gt;     group             0.122    0.020    6.036    0.000
#&gt;   zP38 ~                                              
#&gt;     group            -0.062    0.019   -3.274    0.001
#&gt;   zPIP2 ~                                             
#&gt;     group             0.110    0.021    5.243    0.000
#&gt;   zPIP3 ~                                             
#&gt;     group            -0.301    0.023  -13.278    0.000
#&gt;   zPKA ~                                              
#&gt;     group             0.290    0.023   12.720    0.000
#&gt;   zPKC ~                                              
#&gt;     group             0.116    0.024    4.818    0.000
#&gt;   zPlcg ~                                             
#&gt;     group             0.268    0.024   11.359    0.000
#&gt;   zRaf ~                                              
#&gt;     group            -0.467    0.022  -21.080    0.000
#&gt;   zAkt ~                                              
#&gt;     zPIP3            -0.023    0.020   -1.148    0.251
#&gt;     zPKA              0.507    0.020   24.859    0.000
#&gt;   zErk ~                                              
#&gt;     zMek              0.016    0.021    0.748    0.454
#&gt;     zPKA              0.407    0.022   18.760    0.000
#&gt;   zJnk ~                                              
#&gt;     zPKA             -0.002    0.023   -0.080    0.936
#&gt;     zPKC              0.096    0.022    4.284    0.000
#&gt;   zMek ~                                              
#&gt;     zPKA             -0.017    0.018   -0.961    0.337
#&gt;     zPKC             -0.015    0.017   -0.886    0.376
#&gt;     zRaf              0.736    0.019   37.861    0.000
#&gt;   zP38 ~                                              
#&gt;     zPKA             -0.000    0.019   -0.017    0.986
#&gt;     zPKC              0.652    0.018   35.760    0.000
#&gt;   zPIP2 ~                                             
#&gt;     zPIP3             0.528    0.021   25.141    0.000
#&gt;     zPlcg             0.199    0.020    9.762    0.000
#&gt;   zPKC ~                                              
#&gt;     zPIP2             0.029    0.025    1.163    0.245
#&gt;     zPlcg            -0.010    0.026   -0.383    0.702
#&gt;   zPlcg ~                                             
#&gt;     zPIP3             0.288    0.024   12.208    0.000
#&gt;   zRaf ~                                              
#&gt;     zPKA              0.008    0.022    0.365    0.715
#&gt;     zPKC              0.007    0.021    0.348    0.728
#&gt; 
#&gt; Covariances:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;  .zPIP3 ~~                                            
#&gt;    .zPKA             -0.016    0.022   -0.724    0.469
#&gt; 
#&gt; Variances:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;    .zAkt              0.671    0.023   29.715    0.000
#&gt;    .zErk              0.762    0.026   29.715    0.000
#&gt;    .zJnk              0.872    0.029   29.715    0.000
#&gt;    .zMek              0.523    0.018   29.715    0.000
#&gt;    .zP38              0.579    0.019   29.715    0.000
#&gt;    .zPIP2             0.653    0.022   29.715    0.000
#&gt;    .zPIP3             0.909    0.031   29.715    0.000
#&gt;    .zPKA              0.916    0.031   29.715    0.000
#&gt;    .zPKC              0.986    0.033   29.715    0.000
#&gt;    .zPlcg             0.891    0.030   29.715    0.000
#&gt;    .zRaf              0.785    0.026   29.715    0.000
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>sem1</span><span class='op'>$</span><span class='va'>gest</span><span class='op'>)</span>
</div><div class='output co'>#&gt;     lhs op   rhs    est    se       z pvalue ci.lower ci.upper
#&gt; 1   Akt  ~ group  0.150 0.021   7.057  0.000    0.108    0.192
#&gt; 2   Erk  ~ group  0.179 0.022   8.046  0.000    0.135    0.222
#&gt; 3   Jnk  ~ group  0.333 0.023  14.279  0.000    0.288    0.379
#&gt; 4   Mek  ~ group  0.122 0.020   6.036  0.000    0.082    0.162
#&gt; 5   P38  ~ group -0.062 0.019  -3.274  0.001   -0.100   -0.025
#&gt; 6  PIP2  ~ group  0.110 0.021   5.243  0.000    0.069    0.150
#&gt; 7  PIP3  ~ group -0.301 0.023 -13.278  0.000   -0.346   -0.257
#&gt; 8   PKA  ~ group  0.290 0.023  12.720  0.000    0.245    0.334
#&gt; 9   PKC  ~ group  0.116 0.024   4.818  0.000    0.069    0.163
#&gt; 10 Plcg  ~ group  0.268 0.024  11.359  0.000    0.222    0.314
#&gt; 11  Raf  ~ group -0.467 0.022 -21.080  0.000   -0.510   -0.423</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'>parameterEstimates</span><span class='op'>(</span><span class='va'>sem1</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;     lhs op   rhs    est    se      z pvalue ci.lower ci.upper
#&gt; 1  zAkt  ~ group  0.150 0.021  7.057  0.000    0.108    0.192
#&gt; 2  zErk  ~ group  0.179 0.022  8.046  0.000    0.135    0.222
#&gt; 3  zJnk  ~ group  0.333 0.023 14.279  0.000    0.288    0.379
#&gt; 4  zMek  ~ group  0.122 0.020  6.036  0.000    0.082    0.162
#&gt; 5  zP38  ~ group -0.062 0.019 -3.274  0.001   -0.100   -0.025
#&gt; 6 zPIP2  ~ group  0.110 0.021  5.243  0.000    0.069    0.150</div><div class='input'>
<span class='co'># Graphs</span>
<span class='fu'><a href='gplot.html'>gplot</a></span><span class='op'>(</span><span class='va'>sem1</span><span class='op'>$</span><span class='va'>graph</span>, main <span class='op'>=</span> <span class='st'>"node differences"</span><span class='op'>)</span>
</div><div class='img'><img src='SEMrun-3.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>sem1</span><span class='op'>$</span><span class='va'>graph</span>, layout <span class='op'>=</span> <span class='va'>layout.circle</span>, main <span class='op'>=</span> <span class='st'>"node differences"</span><span class='op'>)</span>
</div><div class='img'><img src='SEMrun-4.png' alt='' width='700' height='433' /></div><div class='input'>

<span class='co'>#### Two-group model fitting (group effect on edges)</span>

<span class='va'>sem2</span> <span class='op'>&lt;-</span> <span class='fu'>SEMrun</span><span class='op'>(</span>graph <span class='op'>=</span> <span class='va'>sachs</span><span class='op'>$</span><span class='va'>graph</span>, data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log</a></span><span class='op'>(</span><span class='va'>sachs</span><span class='op'>$</span><span class='va'>pkc</span><span class='op'>)</span>,
               group <span class='op'>=</span> <span class='va'>sachs</span><span class='op'>$</span><span class='va'>group</span>,
               fit <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: lavaan WARNING:</span>
#&gt; <span class='warning'>    model syntax contains variance/covariance/intercept formulas</span>
#&gt; <span class='warning'>    involving (an) exogenous variable(s): [zPIP3 zPKA]; These</span>
#&gt; <span class='warning'>    variables will now be treated as random introducing additional</span>
#&gt; <span class='warning'>    free parameters. If you wish to treat those variables as fixed,</span>
#&gt; <span class='warning'>    remove these formulas from the model syntax. Otherwise, consider</span>
#&gt; <span class='warning'>    adding the fixed.x = FALSE option.</span></div><div class='output co'>#&gt; NLMINB solver ended normally after 17 iterations 
#&gt; 
#&gt; deviance/df: 31.1974  srmr: 0.0815954 
#&gt; 
#&gt; Brown's combined P-value of edge activation: 0 
#&gt; 
#&gt; Brown's combined P-value of edge inhibition: 0.9995853 
#&gt; </div><div class='input'>
<span class='co'># Summaries</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>sem2</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>)</span>
</div><div class='output co'>#&gt; lavaan 0.6-8 ended normally after 17 iterations
#&gt; 
#&gt;   Estimator                                         ML
#&gt;   Optimization method                           NLMINB
#&gt;   Number of model parameters                        60
#&gt;                                                       
#&gt;   Number of observations per group:                   
#&gt;     Group 1                                        853
#&gt;     Group 2                                        913
#&gt;                                                       
#&gt; Model Test User Model:
#&gt;                                                       
#&gt;   Test statistic                              2246.212
#&gt;   Degrees of freedom                                72
#&gt;   P-value (Chi-square)                           0.000
#&gt;   Test statistic for each group:
#&gt;     Group 1                                    877.592
#&gt;     Group 2                                   1368.621
#&gt; 
#&gt; Parameter Estimates:
#&gt; 
#&gt;   Standard errors                             Standard
#&gt;   Information                                 Observed
#&gt;   Observed information based on                Hessian
#&gt; 
#&gt; 
#&gt; Group 1 [Group 1]:
#&gt; 
#&gt; Regressions:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;   zAkt ~                                              
#&gt;     zPIP3            -0.035    0.031   -1.129    0.259
#&gt;     zPKA              0.404    0.031   12.908    0.000
#&gt;   zErk ~                                              
#&gt;     zMek              0.026    0.032    0.804    0.422
#&gt;     zPKA              0.351    0.032   10.936    0.000
#&gt;   zJnk ~                                              
#&gt;     zPKA             -0.003    0.034   -0.091    0.927
#&gt;     zPKC             -0.202    0.034   -6.021    0.000
#&gt;   zMek ~                                              
#&gt;     zPKA             -0.002    0.025   -0.081    0.936
#&gt;     zPKC             -0.019    0.025   -0.756    0.450
#&gt;     zRaf              0.678    0.025   26.971    0.000
#&gt;   zP38 ~                                              
#&gt;     zPKA              0.008    0.028    0.286    0.775
#&gt;     zPKC              0.581    0.028   20.850    0.000
#&gt;   zPIP2 ~                                             
#&gt;     zPIP3             0.350    0.032   10.854    0.000
#&gt;     zPlcg             0.003    0.032    0.092    0.927
#&gt;   zPKC ~                                              
#&gt;     zPIP2            -0.017    0.034   -0.511    0.610
#&gt;     zPlcg            -0.016    0.034   -0.455    0.649
#&gt;   zPlcg ~                                             
#&gt;     zPIP3             0.090    0.034    2.653    0.008
#&gt;   zRaf ~                                              
#&gt;     zPKA             -0.003    0.034   -0.097    0.923
#&gt;     zPKC             -0.033    0.034   -0.977    0.329
#&gt; 
#&gt; Covariances:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;   zPIP3 ~~                                            
#&gt;     zPKA              0.011    0.034    0.315    0.753
#&gt; 
#&gt; Variances:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;    .zAkt              0.835    0.040   20.652    0.000
#&gt;    .zErk              0.876    0.042   20.652    0.000
#&gt;    .zJnk              0.958    0.046   20.652    0.000
#&gt;    .zMek              0.538    0.026   20.652    0.000
#&gt;    .zP38              0.661    0.032   20.652    0.000
#&gt;    .zPIP2             0.877    0.042   20.652    0.000
#&gt;    .zPKC              0.998    0.048   20.652    0.000
#&gt;    .zPlcg             0.991    0.048   20.652    0.000
#&gt;    .zRaf              0.998    0.048   20.652    0.000
#&gt;     zPIP3             0.999    0.048   20.652    0.000
#&gt;     zPKA              0.999    0.048   20.652    0.000
#&gt; 
#&gt; 
#&gt; Group 2 [Group 2]:
#&gt; 
#&gt; Regressions:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;   zAkt ~                                              
#&gt;     zPIP3            -0.007    0.026   -0.271    0.786
#&gt;     zPKA              0.616    0.026   23.631    0.000
#&gt;   zErk ~                                              
#&gt;     zMek              0.011    0.029    0.391    0.696
#&gt;     zPKA              0.469    0.029   16.021    0.000
#&gt;   zJnk ~                                              
#&gt;     zPKA              0.020    0.028    0.713    0.476
#&gt;     zPKC              0.523    0.028   18.558    0.000
#&gt;   zMek ~                                              
#&gt;     zPKA             -0.032    0.025   -1.278    0.201
#&gt;     zPKC             -0.016    0.025   -0.634    0.526
#&gt;     zRaf              0.665    0.025   26.816    0.000
#&gt;   zP38 ~                                              
#&gt;     zPKA              0.001    0.023    0.027    0.978
#&gt;     zPKC              0.718    0.023   31.170    0.000
#&gt;   zPIP2 ~                                             
#&gt;     zPIP3             0.545    0.024   22.516    0.000
#&gt;     zPlcg             0.332    0.024   13.691    0.000
#&gt;   zPKC ~                                              
#&gt;     zPIP2             0.078    0.040    1.963    0.050
#&gt;     zPlcg            -0.027    0.040   -0.683    0.494
#&gt;   zPlcg ~                                             
#&gt;     zPIP3             0.424    0.030   14.145    0.000
#&gt;   zRaf ~                                              
#&gt;     zPKA              0.023    0.033    0.694    0.488
#&gt;     zPKC              0.049    0.033    1.493    0.136
#&gt; 
#&gt; Covariances:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;   zPIP3 ~~                                            
#&gt;     zPKA             -0.042    0.033   -1.275    0.202
#&gt; 
#&gt; Variances:
#&gt;                    Estimate  Std.Err  z-value  P(&gt;|z|)
#&gt;    .zAkt              0.619    0.029   21.366    0.000
#&gt;    .zErk              0.780    0.036   21.366    0.000
#&gt;    .zJnk              0.725    0.034   21.366    0.000
#&gt;    .zMek              0.558    0.026   21.366    0.000
#&gt;    .zP38              0.484    0.023   21.366    0.000
#&gt;    .zPIP2             0.439    0.021   21.366    0.000
#&gt;    .zPKC              0.994    0.047   21.366    0.000
#&gt;    .zPlcg             0.819    0.038   21.366    0.000
#&gt;    .zRaf              0.996    0.047   21.366    0.000
#&gt;     zPIP3             0.999    0.047   21.366    0.000
#&gt;     zPKA              0.999    0.047   21.366    0.000
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>sem2</span><span class='op'>$</span><span class='va'>dest</span><span class='op'>)</span>
</div><div class='output co'>#&gt;     lhs op  rhs  d_est  d_se    d_z pvalue d_lower d_upper
#&gt; 1   Akt  ~ PIP3  0.028 0.041  0.693  0.488  -0.052   0.108
#&gt; 2   Akt  ~  PKA  0.212 0.041  5.208  0.000   0.132   0.292
#&gt; 3   Erk  ~  Mek -0.014 0.043 -0.330  0.741  -0.099   0.071
#&gt; 4   Erk  ~  PKA  0.118 0.043  2.718  0.007   0.033   0.203
#&gt; 5   Jnk  ~  PKA  0.023 0.044  0.528  0.598  -0.063   0.109
#&gt; 6   Jnk  ~  PKC  0.725 0.044 16.544  0.000   0.639   0.811
#&gt; 7   Mek  ~  PKA -0.030 0.035 -0.839  0.402  -0.099   0.040
#&gt; 8   Mek  ~  PKC  0.003 0.035  0.094  0.925  -0.066   0.073
#&gt; 9   Mek  ~  Raf -0.014 0.035 -0.386  0.699  -0.083   0.056
#&gt; 10  P38  ~  PKA -0.007 0.036 -0.203  0.839  -0.078   0.064
#&gt; 11  P38  ~  PKC  0.137 0.036  3.777  0.000   0.066   0.208
#&gt; 12 PIP2  ~ PIP3  0.196 0.040  4.858  0.000   0.117   0.275
#&gt; 13 PIP2  ~ Plcg  0.329 0.040  8.155  0.000   0.250   0.408
#&gt; 14  PKC  ~ PIP2  0.096 0.053  1.823  0.068  -0.007   0.199
#&gt; 15  PKC  ~ Plcg -0.012 0.053 -0.223  0.824  -0.115   0.091
#&gt; 16 Plcg  ~ PIP3  0.334 0.045  7.346  0.000   0.245   0.422
#&gt; 17  Raf  ~  PKA  0.026 0.048  0.551  0.581  -0.067   0.120
#&gt; 18  Raf  ~  PKC  0.083 0.048  1.739  0.082  -0.011   0.176</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'>parameterEstimates</span><span class='op'>(</span><span class='va'>sem2</span><span class='op'>$</span><span class='va'>fit</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;    lhs op   rhs block group    est    se      z pvalue ci.lower ci.upper
#&gt; 1 zAkt  ~ zPIP3     1     1 -0.035 0.031 -1.129  0.259   -0.097    0.026
#&gt; 2 zAkt  ~  zPKA     1     1  0.404 0.031 12.908  0.000    0.343    0.465
#&gt; 3 zErk  ~  zMek     1     1  0.026 0.032  0.804  0.422   -0.037    0.089
#&gt; 4 zErk  ~  zPKA     1     1  0.351 0.032 10.936  0.000    0.288    0.413
#&gt; 5 zJnk  ~  zPKA     1     1 -0.003 0.034 -0.091  0.927   -0.069    0.063
#&gt; 6 zJnk  ~  zPKC     1     1 -0.202 0.034 -6.021  0.000   -0.268   -0.136</div><div class='input'>
<span class='co'># Graphs</span>
<span class='fu'><a href='gplot.html'>gplot</a></span><span class='op'>(</span><span class='va'>sem2</span><span class='op'>$</span><span class='va'>graph</span>, main <span class='op'>=</span> <span class='st'>"Between group edge differences"</span><span class='op'>)</span>
</div><div class='img'><img src='SEMrun-5.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>sem2</span><span class='op'>$</span><span class='va'>graph</span>, layout <span class='op'>=</span> <span class='va'>layout.circle</span>, main <span class='op'>=</span> <span class='st'>"Between group edge differences"</span><span class='op'>)</span>
</div><div class='img'><img src='SEMrun-6.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># \donttest{</span>

<span class='co'># Fitting and visualization of a large pathway:</span>

<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>graphite</span><span class='op'>)</span>
<span class='va'>humanKegg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/graphite/man/pathways.html'>pathways</a></span><span class='op'>(</span><span class='st'>"hsapiens"</span>, <span class='st'>"kegg"</span><span class='op'>)</span>
<span class='va'>p</span> <span class='op'>&lt;-</span> <span class='va'>humanKegg</span><span class='op'>[[</span><span class='st'>"MAPK signaling pathway"</span><span class='op'>]</span><span class='op'>]</span>
<span class='va'>g</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/graphite/man/pathwayGraph.html'>pathwayGraph</a></span><span class='op'>(</span><span class='va'>p</span><span class='op'>)</span>
<span class='fu'>graph</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/graph/man/graphNEL-class.html'>nodes</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>gsub</a></span><span class='op'>(</span><span class='st'>"ENTREZID:"</span>,<span class='st'>""</span>,<span class='fu'>graph</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/graph/man/graphNEL-class.html'>nodes</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>G</span> <span class='op'>&lt;-</span> <span class='fu'><a href='properties.html'>properties</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/igraph/man/graph_from_graphnel.html'>graph_from_graphnel</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>)</span><span class='op'>)</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
</div><div class='output co'>#&gt; Frequency distribution of graph components
#&gt; 
#&gt;   n.nodes n.graphs
#&gt; 1       3        1
#&gt; 2     291        1
#&gt; 
#&gt; Percent of vertices in the giant component: 99 %
#&gt; 
#&gt;   is.simple      is.dag is.directed is.weighted 
#&gt;        TRUE       FALSE        TRUE        TRUE 
#&gt; 
#&gt; which.mutual.FALSE  which.mutual.TRUE 
#&gt;               1965                  6 </div><div class='input'>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>huge</span><span class='op'>)</span>
<span class='va'>als.npn</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/huge/man/huge.npn.html'>huge.npn</a></span><span class='op'>(</span><span class='va'>alsData</span><span class='op'>$</span><span class='va'>exprs</span><span class='op'>)</span>
</div><div class='output co'>#&gt; Conducting the nonparanormal (npn) transformation via shrunkun ECDF....done.</div><div class='input'>
<span class='va'>g1</span> <span class='op'>&lt;-</span> <span class='fu'>SEMrun</span><span class='op'>(</span><span class='va'>G</span>, <span class='va'>als.npn</span>, <span class='va'>alsData</span><span class='op'>$</span><span class='va'>group</span>, algo <span class='op'>=</span> <span class='st'>"cggm"</span><span class='op'>)</span><span class='op'>$</span><span class='va'>graph</span>
</div><div class='output co'>#&gt; Estimating optimal shrinkage intensity lambda (correlation matrix): 0.0647 
#&gt; 
#&gt; GGM (constrained) solver ended normally after 0 iterations 
#&gt; 
#&gt; deviance/df: 1.597313  srmr: 0.309441 
#&gt; 
#&gt; Brown's combined P-value of node activation: 0 
#&gt; 
#&gt; Brown's combined P-value of node inhibition: 0 
#&gt; </div><div class='input'><span class='va'>g2</span> <span class='op'>&lt;-</span> <span class='fu'>SEMrun</span><span class='op'>(</span><span class='va'>g1</span>, <span class='va'>als.npn</span>, <span class='va'>alsData</span><span class='op'>$</span><span class='va'>group</span>, fit <span class='op'>=</span> <span class='fl'>2</span>, algo <span class='op'>=</span> <span class='st'>"cggm"</span><span class='op'>)</span><span class='op'>$</span><span class='va'>graph</span>
</div><div class='output co'>#&gt; GGM (constrained) solver ended normally after 0 iterations 
#&gt; 
#&gt; deviance/df: 0.7447792  srmr: 0.3063258 
#&gt; 
#&gt; Brown's combined P-value of edge activation: 0 
#&gt; 
#&gt; Brown's combined P-value of edge inhibition: 1.525516e-08 
#&gt; </div><div class='input'>
<span class='co'># extract the subgraph with between group node and edge differences</span>
<span class='va'>g2</span> <span class='op'>&lt;-</span> <span class='va'>g2</span> <span class='op'>-</span> <span class='fu'><a href='https://rdrr.io/pkg/igraph/man/E.html'>E</a></span><span class='op'>(</span><span class='va'>g2</span><span class='op'>)</span><span class='op'>[</span><span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/igraph/man/E.html'>E</a></span><span class='op'>(</span><span class='va'>g2</span><span class='op'>)</span><span class='op'>$</span><span class='va'>color</span> <span class='op'>!=</span> <span class='st'>"gray50"</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>g</span> <span class='op'>&lt;-</span> <span class='fu'><a href='properties.html'>properties</a></span><span class='op'>(</span><span class='va'>g2</span><span class='op'>)</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
</div><div class='output co'>#&gt; Frequency distribution of graph components
#&gt; 
#&gt;   n.nodes n.graphs
#&gt; 1       1      134
#&gt; 2       2        5
#&gt; 3       3        1
#&gt; 4       4        1
#&gt; 5      13        1
#&gt; 6      32        1
#&gt; 7      71        1
#&gt; 
#&gt; Percent of vertices in the giant component: 26.6 %
#&gt; 
#&gt;   is.simple      is.dag is.directed is.weighted 
#&gt;        TRUE        TRUE        TRUE        TRUE 
#&gt; 
#&gt; which.mutual.FALSE 
#&gt;                 82 </div><div class='input'>
<span class='co'># plot graph</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>org.Hs.eg.db</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: AnnotationDbi</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocGenerics</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: parallel</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'BiocGenerics'</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:parallel':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
#&gt; <span class='message'>    clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
#&gt; <span class='message'>    parLapplyLB, parRapply, parSapply, parSapplyLB</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:igraph':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    normalize, path, union</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:stats':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    IQR, mad, sd, var, xtabs</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:base':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
#&gt; <span class='message'>    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
#&gt; <span class='message'>    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
#&gt; <span class='message'>    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
#&gt; <span class='message'>    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
#&gt; <span class='message'>    union, unique, unsplit, which.max, which.min</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'S4Vectors'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:base':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'IRanges'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:grDevices':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    windows</span></div><div class='output co'>#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/pkg/igraph/man/V.html'>V</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>)</span><span class='op'>$</span><span class='va'>label</span> <span class='op'>&lt;-</span> <span class='fu'>mapIds</span><span class='op'>(</span><span class='va'>org.Hs.eg.db</span>, <span class='fu'><a href='https://rdrr.io/pkg/igraph/man/V.html'>V</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>)</span><span class='op'>$</span><span class='va'>name</span>, <span class='st'>'SYMBOL'</span>, <span class='st'>'ENTREZID'</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>'select()' returned 1:1 mapping between keys and columns</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/pkg/igraph/man/E.html'>E</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>)</span><span class='op'>$</span><span class='va'>color</span><span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/igraph/man/E.html'>E</a></span><span class='op'>(</span><span class='va'>g2</span><span class='op'>)</span><span class='op'>$</span><span class='va'>color</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/pkg/igraph/man/E.html'>E</a></span><span class='op'>(</span><span class='va'>g2</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/pkg/igraph/man/E.html'>E</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>)</span><span class='op'>]</span>
<span class='fu'><a href='gplot.html'>gplot</a></span><span class='op'>(</span><span class='va'>g</span>, l <span class='op'>=</span> <span class='st'>"fdp"</span>, main<span class='op'>=</span><span class='st'>"node and edge group differences"</span><span class='op'>)</span>
</div><div class='img'><img src='SEMrun-7.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># }</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Mario Grassi, Fernando Palluzzi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


