<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Fit a graph as a Structural Equation Model (SEM) — SEMrun • SEMgraph</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Fit a graph as a Structural Equation Model (SEM) — SEMrun"><meta property="og:description" content="SEMrun() converts a (directed, undirected, or mixed)
graph to a SEM and fits it. If a binary group variable (i.e., case/control)
is present, node-level or edge-level perturbation is evaluated.
This function can handle loop-containing models, although multiple
links between the same two nodes (including self-loops and mutual
interactions) and bows (i.e., a directed and a bidirected link between
two nodes) are not allowed."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SEMgraph</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/fernandoPalluzzi/SEMgraph/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit a graph as a Structural Equation Model (SEM)</h1>
    <small class="dont-index">Source: <a href="https://github.com/fernandoPalluzzi/SEMgraph/blob/HEAD/R/semFit.R" class="external-link"><code>R/semFit.R</code></a></small>
    <div class="hidden name"><code>SEMrun.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>SEMrun()</code> converts a (directed, undirected, or mixed)
graph to a SEM and fits it. If a binary group variable (i.e., case/control)
is present, node-level or edge-level perturbation is evaluated.
This function can handle loop-containing models, although multiple
links between the same two nodes (including self-loops and mutual
interactions) and bows (i.e., a directed and a bidirected link between
two nodes) are not allowed.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">SEMrun</span><span class="op">(</span>
  <span class="va">graph</span>,
  <span class="va">data</span>,
  group <span class="op">=</span> <span class="cn">NULL</span>,
  fit <span class="op">=</span> <span class="fl">0</span>,
  algo <span class="op">=</span> <span class="st">"lavaan"</span>,
  start <span class="op">=</span> <span class="cn">NULL</span>,
  SE <span class="op">=</span> <span class="st">"standard"</span>,
  n_rep <span class="op">=</span> <span class="fl">1000</span>,
  limit <span class="op">=</span> <span class="fl">100</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>graph</dt>
<dd><p>An igraph object.</p></dd>


<dt>data</dt>
<dd><p>A matrix whith rows corresponding to subjects, and
columns to graph nodes (variables).</p></dd>


<dt>group</dt>
<dd><p>A binary vector. This vector must be as long as the
number of subjects. Each vector element must be 1 for cases and 0
for control subjects. If <code>NULL</code> (default), group influence will
not be considered.</p></dd>


<dt>fit</dt>
<dd><p>A numeric value indicating the SEM fitting mode.
If <code>fit = 0</code> (default), no group effect is considered.
If <code>fit = 1</code>, a "common" model is used to evaluate group effects
on graph nodes.
If <code>fit = 2</code>, a two-group model is used to evaluate group effects
on graph edges.</p></dd>


<dt>algo</dt>
<dd><p>MLE method used for SEM fitting. If <code>algo = "lavaan"</code>
(default), the SEM will be fitted using the NLMINB solver from
<code>lavaan</code> R package, with standard errors derived from the expected
Fisher information matrix. If <code>algo = "ricf"</code>, the model is fitted
via residual iterative conditional fitting (RICF; Drton et al. 2009).
If <code>algo = "cggm"</code>, model fitting is based on constrained Gaussian
Graphical Modeling (GGM) and de-sparsified glasso estimator
(Williams, 2020).</p></dd>


<dt>start</dt>
<dd><p>Starting value of SEM parameters for <code>algo = "lavaan"</code>.
If start is <code>NULL</code> (default), the algorithm will determine the
starting values. If start is a numeric value, it will be used as a
scaling factor for the edge weights in the graph object (graph attribute
<code>E(graph)$weight</code>).
For instance, a scaling factor is useful when weights have fixed values
(e.g., 1 for activated, -1 for repressed, and 0 for unchanged interaction).
Fixed values may compromise model fitting, and scaling them is a safe
option to avoid this problem. As a rule of thumb, to our experience,
<code>start = 0.1</code> generally performs well with -1, 0, 1 weights.</p></dd>


<dt>SE</dt>
<dd><p>If "standard" (default), with <code>algo = "lavaan"</code>,
conventional standard errors are computed based on inverting the observed
information matrix. If "none", no standard errors are computed.</p></dd>


<dt>n_rep</dt>
<dd><p>Number of randomization replicates (default = 1000),
for permutation flip or boostrap samples, if <code>algo = "ricf"</code>.</p></dd>


<dt>limit</dt>
<dd><p>An integer value corresponding to the network size
(i.e., number of nodes). Beyond this limit, the execution under
<code>algo = "lavaan"</code> will run with <code>SE = "none"</code>, if 
<code>fit = 0</code>, or will be ridirected to <code>algo = "ricf"</code>, if
<code>fit = 1</code>, or to <code>algo = "ggm"</code>, if <code>fit = 2</code>.
This redirection is necessary to reduce the computational demand of
standard error estimation by lavaan. Increasing this number will
enforce lavaan execution when <code>algo = "lavaan"</code>.</p></dd>


<dt>...</dt>
<dd><p>Currently ignored.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of 5 objects:</p><ol><li><p>"fit", SEM fitted lavaan, ricf, or ggmncv object,
depending on the MLE method specified by the <code>algo</code> argument;</p></li>
<li><p>"gest" or "dest", a data.frame of node-specific
("gest") or edge-specific ("dest") group effect estimates and P-values;</p></li>
<li><p>"model", SEM model as a string if <code>algo = "lavaan"</code>,
and <code>NULL</code> otherwise;</p></li>
<li><p>"graph", the induced subgraph of the input network mapped
on data variables. Graph edges (i.e., direct effects) with P-value &lt; 0.05
will be highlighted in red (beta &gt; 0) or blue (beta &lt; 0). If a group
vector is given, nodes with significant group effect (P-value &lt; 0.05)
will be red-shaded (beta &gt; 0) or lightblue-shaded (beta &lt; 0);</p></li>
<li><p>"dataXY", input data subset mapping graph nodes, plus
group at the first column (if no group is specified, this column will
take NA values).</p></li>
</ol></div>
    <div id="details">
    <h2>Details</h2>
    <p>SEMrun maps data onto the input graph and converts it into a
SEM. Directed connections (X -&gt; Y) are interpreted as direct causal
effects, while undirected, mutual, and bidirected connections are
converted into model covariances. SEMrun output contains different sets
of parameter estimates. Beta coefficients (i.e., direct effects) are
estimated from directed interactions and residual covariances (psi
coefficients) from bidirected, undirected, or mutual interactions.
If a group variable is given, exogenous group effects on nodes (gamma
coefficients) will be estimated. This will also lead to the estimation
of a set of aggregated group effects, if <code>algo = "ricf"</code> (see
<code><a href="SEMgsa.html">SEMgsa</a></code>).
By default, maximum likelihood parameter estimates and P-values for
parameter sets are computed by conventional z-test (= estimate/SE),
and fits it through the <code><a href="https://rdrr.io/pkg/lavaan/man/lavaan.html" class="external-link">lavaan</a></code> function, via
Maximum Likelihood Estimation (estimator = "ML", default estimator in
<code><a href="https://rdrr.io/pkg/lavaan/man/lavOptions.html" class="external-link">lavOptions</a></code>).
In case of high dimensionality (n.variables &gt;&gt; n.subjects), the covariance
matrix could not be semi-definite positive and thus parameter estimates
could not be done. If this happens, covariance matrix regularization
is enabled using the James-Stein-type shrinkage estimator implemented
in the function <code><a href="https://rdrr.io/pkg/corpcor/man/pcor.shrink.html" class="external-link">pcor.shrink</a></code> of corpcor R package.
Argument <code>fit</code> determines how group influence is evaluated in the
model, as absent (<code>fit = 0</code>), node perturbation (<code>fit = 1</code>),
or edge perturbation (<code>fit = 2</code>). When <code>fit = 1</code>, the group
is modeled as an exogenous variable, influencing all the other graph
nodes. When <code>fit = 2</code>, SEMrun estimates the differences
of the beta and/or psi coefficients (network edges) between groups.
This is equivalent to fit a separate model for cases and controls,
as opposed to one common model perturbed by the exogenous group effect.
Once fitted, the two models are then compared to assess significant
edge (i.e., direct effect) differences (d = beta1 - beta0).
P-values for parameter sets are computed by z-test (= d/SE), through
<code><a href="https://rdrr.io/pkg/lavaan/man/lavaan.html" class="external-link">lavaan</a></code>. As an alternative to standard P-value
calculation, SEMrun may use either RICF (randomization P-values) or
GGM (de-sparsified P-values) methods. These algorithms are much faster
than <code><a href="https://rdrr.io/pkg/lavaan/man/lavaan.html" class="external-link">lavaan</a></code> in case of large input graphs.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Pearl J (1998). Graphs, Causality, and Structural Equation Models.
Sociological Methods &amp; Research., 27(2):226-284.
&lt;https://doi.org/10.1177/0049124198027002004&gt;</p>
<p>Yves Rosseel (2012). lavaan: An R Package for Structural Equation
Modeling. Journal of Statistical Software, 48(2): 1-36.
&lt;https://www.jstatsoft.org/v48/i02/&gt;</p>
<p>Pepe D, Grassi M (2014). Investigating perturbed pathway modules
from gene expression data via Structural Equation Models. BMC
Bioinformatics, 15: 132.
&lt;https://doi.org/10.1186/1471-2105-15-132&gt;</p>
<p>Drton M, Eichler M, Richardson TS (2009). Computing Maximum Likelihood
Estimated in Recursive Linear Models with Correlated Errors.
Journal of Machine Learning Research, 10(Oct): 2329-2348.
&lt;https://www.jmlr.org/papers/volume10/drton09a/drton09a.pdf&gt;</p>
<p>Larson JL and Owen AB (2015). Moment based gene set tests. BMC
Bioinformatics, 16: 132. &lt;https://doi.org/10.1186/s12859-015-0571-7&gt;</p>
<p>Palluzzi F, Grassi M (2021). SEMgraph: An R Package for Causal Network
Analysis of High-Throughput Data with Structural Equation Models.
&lt;arXiv:2103.08332&gt;</p>
<p>Williams D (2020). GGMncv: Gaussian Graphical Models with Non-Convex
Penalties. R package version 1.1.0.
&lt;https://CRAN.R-project.org/package=GGMncv/&gt;</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>See <code><a href="https://rdrr.io/pkg/ggm/man/fitAncestralGraph.html" class="external-link">fitAncestralGraph</a></code> for RICF algorithm
details, <code><a href="https://rdrr.io/pkg/flip/man/flip.html" class="external-link">flip</a></code> for randomization P-values, and
<code><a href="https://rdrr.io/pkg/GGMncv/man/constrained.html" class="external-link">constrained</a></code> for constrained GGM, and
<code><a href="https://rdrr.io/pkg/GGMncv/man/inference.html" class="external-link">inference</a></code> for de-sparsified P-values.</p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Mario Grassi <a href="mailto:mario.grassi@unipv.it">mario.grassi@unipv.it</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="co">#### Model fitting (no group effect)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">sem0</span> <span class="op">&lt;-</span> <span class="fu">SEMrun</span><span class="op">(</span>graph <span class="op">=</span> <span class="va">sachs</span><span class="op">$</span><span class="va">graph</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">sachs</span><span class="op">$</span><span class="va">pkc</span><span class="op">)</span>, algo <span class="op">=</span> <span class="st">"lavaan"</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NLMINB solver ended normally after 10 iterations </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> deviance/df: 66.93929  srmr: 0.08947309 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">sem0</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> lavaan 0.6-11 ended normally after 10 iterations</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Estimator                                         ML</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Optimization method                           NLMINB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of model parameters                        30</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of observations                          1766</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Model Test User Model:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Test statistic                              2409.815</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Degrees of freedom                                36</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   P-value (Chi-square)                           0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Parameter Estimates:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Standard errors                             Standard</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Information                                 Observed</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Observed information based on                Hessian</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Regressions:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zAkt ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3            -0.065    0.020   -3.251    0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.546    0.020   27.454    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zErk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zMek             -0.021    0.021   -0.974    0.330</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.456    0.021   21.506    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zJnk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.093    0.024    3.940    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.128    0.024    5.439    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zMek ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.011    0.018    0.615    0.539</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC             -0.006    0.017   -0.321    0.748</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zRaf              0.683    0.018   38.931    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zP38 ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.018    0.018   -0.988    0.323</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.646    0.018   35.504    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP2 ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.489    0.020   24.698    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPlcg             0.227    0.020   11.441    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPKC ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP2             0.020    0.025    0.782    0.434</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPlcg             0.014    0.025    0.570    0.568</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPlcg ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.207    0.023    8.895    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zRaf ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.124    0.024   -5.261    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC             -0.037    0.024   -1.587    0.113</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Covariances:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP3 ~~                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.103    0.024   -4.306    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variances:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zAkt              0.690    0.023   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zErk              0.790    0.027   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zJnk              0.973    0.033   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zMek              0.534    0.018   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zP38              0.583    0.020   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPIP2             0.663    0.022   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPKC              0.999    0.034   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPlcg             0.957    0.032   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zRaf              0.982    0.033   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.999    0.034   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.999    0.034   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">sem0</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   lhs op  rhs    est    se      z pvalue ci.lower ci.upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Akt  ~ PIP3 -0.065 0.020 -3.251  0.001   -0.104   -0.026</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Akt  ~  PKA  0.546 0.020 27.454  0.000    0.507    0.585</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Erk  ~  Mek -0.021 0.021 -0.974  0.330   -0.062    0.021</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 Erk  ~  PKA  0.456 0.021 21.506  0.000    0.415    0.498</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 Jnk  ~  PKA  0.093 0.024  3.940  0.000    0.047    0.139</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 Jnk  ~  PKC  0.128 0.024  5.439  0.000    0.082    0.174</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Graphs</span></span>
<span class="r-in"><span class="fu"><a href="gplot.html">gplot</a></span><span class="op">(</span><span class="va">sem0</span><span class="op">$</span><span class="va">graph</span>, main <span class="op">=</span> <span class="st">"significant edge weights"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="SEMrun-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sem0</span><span class="op">$</span><span class="va">graph</span>, layout <span class="op">=</span> <span class="va">layout.circle</span>, main <span class="op">=</span> <span class="st">"significant edge weights"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="SEMrun-2.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">#### Model fitting (common model, group effect on nodes)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">sem1</span> <span class="op">&lt;-</span> <span class="fu">SEMrun</span><span class="op">(</span>graph <span class="op">=</span> <span class="va">sachs</span><span class="op">$</span><span class="va">graph</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">sachs</span><span class="op">$</span><span class="va">pkc</span><span class="op">)</span>,</span>
<span class="r-in">               group <span class="op">=</span> <span class="va">sachs</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NLMINB solver ended normally after 5 iterations </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> deviance/df: 61.84434  srmr: 0.0700967 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Brown's combined P-value of node activation: 2.664535e-15 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Brown's combined P-value of node inhibition: 0.8332506 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Fitting summaries</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">sem1</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> lavaan 0.6-11 ended normally after 5 iterations</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Estimator                                         ML</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Optimization method                           NLMINB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of model parameters                        41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of observations                          1766</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Model Test User Model:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Test statistic                              2226.396</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Degrees of freedom                                36</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   P-value (Chi-square)                           0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Parameter Estimates:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Standard errors                             Standard</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Information                                 Observed</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Observed information based on                Hessian</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Regressions:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zAkt ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group             0.150    0.021    7.057    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zErk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group             0.179    0.022    8.046    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zJnk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group             0.333    0.023   14.279    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zMek ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group             0.122    0.020    6.036    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zP38 ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group            -0.062    0.019   -3.274    0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP2 ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group             0.110    0.021    5.243    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP3 ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group            -0.301    0.023  -13.278    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPKA ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group             0.290    0.023   12.720    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPKC ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group             0.116    0.024    4.818    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPlcg ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group             0.268    0.024   11.359    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zRaf ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     group            -0.467    0.022  -21.080    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zAkt ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3            -0.023    0.020   -1.148    0.251</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.507    0.020   24.859    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zErk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zMek              0.016    0.021    0.748    0.454</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.407    0.022   18.760    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zJnk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.002    0.023   -0.080    0.936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.096    0.022    4.284    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zMek ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.017    0.018   -0.961    0.337</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC             -0.015    0.017   -0.886    0.376</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zRaf              0.736    0.019   37.861    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zP38 ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.000    0.019   -0.017    0.986</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.652    0.018   35.760    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP2 ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.528    0.021   25.141    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPlcg             0.199    0.020    9.762    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPKC ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP2             0.029    0.025    1.163    0.245</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPlcg            -0.010    0.026   -0.383    0.702</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPlcg ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.288    0.024   12.208    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zRaf ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.008    0.022    0.365    0.715</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.007    0.021    0.348    0.728</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Covariances:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  .zPIP3 ~~                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPKA             -0.016    0.022   -0.724    0.469</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variances:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zAkt              0.671    0.023   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zErk              0.762    0.026   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zJnk              0.872    0.029   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zMek              0.523    0.018   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zP38              0.579    0.019   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPIP2             0.653    0.022   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPIP3             0.909    0.031   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPKA              0.916    0.031   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPKC              0.986    0.033   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPlcg             0.891    0.030   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zRaf              0.785    0.026   29.715    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sem1</span><span class="op">$</span><span class="va">gest</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     lhs op   rhs    est    se       z pvalue ci.lower ci.upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   Akt  ~ group  0.150 0.021   7.057  0.000    0.108    0.192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   Erk  ~ group  0.179 0.022   8.046  0.000    0.135    0.222</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   Jnk  ~ group  0.333 0.023  14.279  0.000    0.288    0.379</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   Mek  ~ group  0.122 0.020   6.036  0.000    0.082    0.162</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   P38  ~ group -0.062 0.019  -3.274  0.001   -0.100   -0.025</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  PIP2  ~ group  0.110 0.021   5.243  0.000    0.069    0.150</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  PIP3  ~ group -0.301 0.023 -13.278  0.000   -0.346   -0.257</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8   PKA  ~ group  0.290 0.023  12.720  0.000    0.245    0.334</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9   PKC  ~ group  0.116 0.024   4.818  0.000    0.069    0.163</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 Plcg  ~ group  0.268 0.024  11.359  0.000    0.222    0.314</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11  Raf  ~ group -0.467 0.022 -21.080  0.000   -0.510   -0.423</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">sem1</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    lhs op   rhs    est    se      z pvalue ci.lower ci.upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  Akt  ~ group  0.150 0.021  7.057  0.000    0.108    0.192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  Erk  ~ group  0.179 0.022  8.046  0.000    0.135    0.222</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  Jnk  ~ group  0.333 0.023 14.279  0.000    0.288    0.379</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  Mek  ~ group  0.122 0.020  6.036  0.000    0.082    0.162</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  P38  ~ group -0.062 0.019 -3.274  0.001   -0.100   -0.025</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 PIP2  ~ group  0.110 0.021  5.243  0.000    0.069    0.150</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Graphs</span></span>
<span class="r-in"><span class="fu"><a href="gplot.html">gplot</a></span><span class="op">(</span><span class="va">sem1</span><span class="op">$</span><span class="va">graph</span>, main <span class="op">=</span> <span class="st">"Between group node differences"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="SEMrun-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sem1</span><span class="op">$</span><span class="va">graph</span>, layout <span class="op">=</span> <span class="va">layout.circle</span>, main <span class="op">=</span> <span class="st">"Between group node differences"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="SEMrun-4.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">#### Two-group model fitting (group effect on edges)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">sem2</span> <span class="op">&lt;-</span> <span class="fu">SEMrun</span><span class="op">(</span>graph <span class="op">=</span> <span class="va">sachs</span><span class="op">$</span><span class="va">graph</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">sachs</span><span class="op">$</span><span class="va">pkc</span><span class="op">)</span>,</span>
<span class="r-in">               group <span class="op">=</span> <span class="va">sachs</span><span class="op">$</span><span class="va">group</span>,</span>
<span class="r-in">               fit <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NLMINB solver ended normally after 13 iterations </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> deviance/df: 31.1974  srmr: 0.08159536 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Brown's combined P-value of edge activation: 3.774758e-15 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Brown's combined P-value of edge inhibition: 0.871612 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Summaries</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">sem2</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> lavaan 0.6-11 ended normally after 13 iterations</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Estimator                                         ML</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Optimization method                           NLMINB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of model parameters                        60</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of observations per group:                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Group 1                                        853</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Group 2                                        913</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Model Test User Model:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Test statistic                              2246.212</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Degrees of freedom                                72</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   P-value (Chi-square)                           0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Test statistic for each group:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Group 1                                    877.592</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Group 2                                   1368.621</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Parameter Estimates:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Standard errors                             Standard</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Information                                 Observed</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Observed information based on                Hessian</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Group 1 [Group 1]:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Regressions:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zAkt ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3            -0.035    0.031   -1.129    0.259</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.404    0.031   12.908    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zErk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zMek              0.026    0.032    0.804    0.422</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.351    0.032   10.936    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zJnk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.003    0.034   -0.091    0.927</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC             -0.202    0.034   -6.021    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zMek ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.002    0.025   -0.081    0.936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC             -0.019    0.025   -0.756    0.450</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zRaf              0.678    0.025   26.971    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zP38 ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.008    0.028    0.286    0.775</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.581    0.028   20.850    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP2 ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.350    0.032   10.854    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPlcg             0.003    0.032    0.092    0.927</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPKC ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP2            -0.017    0.034   -0.511    0.610</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPlcg            -0.016    0.034   -0.455    0.649</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPlcg ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.090    0.034    2.653    0.008</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zRaf ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.003    0.034   -0.097    0.923</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC             -0.033    0.034   -0.977    0.329</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Covariances:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP3 ~~                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.011    0.034    0.315    0.753</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variances:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zAkt              0.835    0.040   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zErk              0.876    0.042   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zJnk              0.958    0.046   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zMek              0.538    0.026   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zP38              0.661    0.032   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPIP2             0.877    0.042   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPKC              0.998    0.048   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPlcg             0.991    0.048   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zRaf              0.998    0.048   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.999    0.048   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.999    0.048   20.652    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Group 2 [Group 2]:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Regressions:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zAkt ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3            -0.007    0.026   -0.271    0.786</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.616    0.026   23.631    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zErk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zMek              0.011    0.029    0.391    0.696</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.469    0.029   16.021    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zJnk ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.020    0.028    0.713    0.476</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.523    0.028   18.558    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zMek ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.032    0.025   -1.278    0.201</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC             -0.016    0.025   -0.634    0.526</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zRaf              0.665    0.025   26.816    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zP38 ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.001    0.023    0.027    0.978</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.718    0.023   31.170    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP2 ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.545    0.024   22.517    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPlcg             0.332    0.024   13.691    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPKC ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP2             0.078    0.040    1.963    0.050</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPlcg            -0.027    0.040   -0.683    0.495</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPlcg ~                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.424    0.030   14.145    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zRaf ~                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.023    0.033    0.694    0.488</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKC              0.049    0.033    1.493    0.136</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Covariances:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zPIP3 ~~                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA             -0.042    0.033   -1.275    0.202</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variances:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate  Std.Err  z-value  P(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zAkt              0.619    0.029   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zErk              0.780    0.036   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zJnk              0.725    0.034   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zMek              0.558    0.026   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zP38              0.484    0.023   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPIP2             0.439    0.021   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPKC              0.994    0.047   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zPlcg             0.819    0.038   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .zRaf              0.996    0.047   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPIP3             0.999    0.047   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     zPKA              0.999    0.047   21.366    0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sem2</span><span class="op">$</span><span class="va">dest</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     lhs op  rhs  d_est  d_se    d_z pvalue d_lower d_upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   Akt  ~ PIP3  0.028 0.041  0.693  0.488  -0.052   0.108</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   Akt  ~  PKA  0.212 0.041  5.208  0.000   0.132   0.292</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   Erk  ~  Mek -0.014 0.043 -0.330  0.741  -0.099   0.071</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   Erk  ~  PKA  0.118 0.043  2.718  0.007   0.033   0.203</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   Jnk  ~  PKA  0.023 0.044  0.528  0.598  -0.063   0.109</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   Jnk  ~  PKC  0.725 0.044 16.544  0.000   0.639   0.811</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7   Mek  ~  PKA -0.030 0.035 -0.839  0.402  -0.099   0.040</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8   Mek  ~  PKC  0.003 0.035  0.094  0.925  -0.066   0.073</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9   Mek  ~  Raf -0.014 0.035 -0.386  0.699  -0.083   0.056</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10  P38  ~  PKA -0.007 0.036 -0.203  0.839  -0.078   0.064</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11  P38  ~  PKC  0.137 0.036  3.777  0.000   0.066   0.208</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 PIP2  ~ PIP3  0.196 0.040  4.858  0.000   0.117   0.275</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 PIP2  ~ Plcg  0.329 0.040  8.155  0.000   0.250   0.408</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14  PKC  ~ PIP2  0.096 0.053  1.823  0.068  -0.007   0.199</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15  PKC  ~ Plcg -0.012 0.053 -0.223  0.824  -0.115   0.091</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 Plcg  ~ PIP3  0.334 0.045  7.346  0.000   0.245   0.422</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17  Raf  ~  PKA  0.026 0.048  0.551  0.581  -0.067   0.120</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18  Raf  ~  PKC  0.083 0.048  1.739  0.082  -0.011   0.176</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="parameterEstimates.html">parameterEstimates</a></span><span class="op">(</span><span class="va">sem2</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   lhs op  rhs    est    se      z pvalue ci.lower ci.upper group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Akt  ~ PIP3 -0.035 0.031 -1.129  0.259   -0.097    0.026     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Akt  ~  PKA  0.404 0.031 12.908  0.000    0.343    0.465     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Erk  ~  Mek  0.026 0.032  0.804  0.422   -0.037    0.089     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 Erk  ~  PKA  0.351 0.032 10.936  0.000    0.288    0.413     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 Jnk  ~  PKA -0.003 0.034 -0.091  0.927   -0.069    0.063     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 Jnk  ~  PKC -0.202 0.034 -6.021  0.000   -0.268   -0.136     1</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Graphs</span></span>
<span class="r-in"><span class="fu"><a href="gplot.html">gplot</a></span><span class="op">(</span><span class="va">sem2</span><span class="op">$</span><span class="va">graph</span>, main <span class="op">=</span> <span class="st">"Between group edge differences"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="SEMrun-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sem2</span><span class="op">$</span><span class="va">graph</span>, layout <span class="op">=</span> <span class="va">layout.circle</span>, main <span class="op">=</span> <span class="st">"Between group edge differences"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="SEMrun-6.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># \donttest{</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Fitting and visualization of a large pathway:</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">kegg.pathways</span><span class="op">[[</span><span class="st">"MAPK signaling pathway"</span><span class="op">]</span><span class="op">]</span></span>
<span class="r-in"><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="properties.html">properties</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>; <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Frequency distribution of graph components</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   n.nodes n.graphs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       3        1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     291        1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Percent of vertices in the giant component: 99 %</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   is.simple      is.dag is.directed is.weighted </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        TRUE       FALSE        TRUE        TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> which.mutual.FALSE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               1967 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IGRAPH a0e4e78 DNW- 291 1967 -- </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> + attr: name (v/c), weight (e/n)</span>
<span class="r-in"></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">huge</span><span class="op">)</span></span>
<span class="r-in"><span class="va">als.npn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/huge/man/huge.npn.html" class="external-link">huge.npn</a></span><span class="op">(</span><span class="va">alsData</span><span class="op">$</span><span class="va">exprs</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conducting the nonparanormal (npn) transformation via shrunkun ECDF....done.</span>
<span class="r-in"></span>
<span class="r-in"><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu">SEMrun</span><span class="op">(</span><span class="va">G</span>, <span class="va">als.npn</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>, algo <span class="op">=</span> <span class="st">"cggm"</span><span class="op">)</span><span class="op">$</span><span class="va">graph</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Estimating optimal shrinkage intensity lambda (correlation matrix): 0.0647 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GGM (constrained) solver ended normally after 0 iterations </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> deviance/df: 1.597429  srmr: 0.2617292 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Brown's combined P-value of node activation: 1.224552e-227 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Brown's combined P-value of node inhibition: 1.354196e-122 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span class="va">g2</span> <span class="op">&lt;-</span> <span class="fu">SEMrun</span><span class="op">(</span><span class="va">g1</span>, <span class="va">als.npn</span>, <span class="va">alsData</span><span class="op">$</span><span class="va">group</span>, fit <span class="op">=</span> <span class="fl">2</span>, algo <span class="op">=</span> <span class="st">"cggm"</span><span class="op">)</span><span class="op">$</span><span class="va">graph</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GGM (constrained) solver ended normally after 0 iterations </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> deviance/df: 0.7447878  srmr: 0.273665 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Brown's combined P-value of edge activation: 2.691819e-22 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Brown's combined P-value of edge inhibition: 4.086134e-23 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># extract the subgraph with between group node and edge differences</span></span>
<span class="r-in"><span class="va">g2</span> <span class="op">&lt;-</span> <span class="va">g2</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">g2</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">g2</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">!=</span> <span class="st">"gray50"</span><span class="op">)</span><span class="op">]</span></span>
<span class="r-in"><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="properties.html">properties</a></span><span class="op">(</span><span class="va">g2</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Frequency distribution of graph components</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   n.nodes n.graphs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       2        5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       3        1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       4        1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      13        1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      32        1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      72        1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Percent of vertices in the giant component: 27 %</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   is.simple      is.dag is.directed is.weighted </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        TRUE        TRUE        TRUE        TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> which.mutual.FALSE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 83 </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># plot graph</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: AnnotationDbi</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from 'package:igraph':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     normalize, path, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from 'package:stats':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from 'package:base':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     tapply, union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: 'S4Vectors'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from 'package:base':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: 'IRanges'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from 'package:grDevices':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     windows</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">'SYMBOL'</span>, <span class="st">'ENTREZID'</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">color</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">g2</span><span class="op">)</span><span class="op">$</span><span class="va">color</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">g2</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">]</span></span>
<span class="r-in"><span class="fu"><a href="gplot.html">gplot</a></span><span class="op">(</span><span class="va">g</span>, l <span class="op">=</span> <span class="st">"fdp"</span>, main<span class="op">=</span><span class="st">"node and edge group differences"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="SEMrun-7.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># }</span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Mario Grassi, Fernando Palluzzi, Barbara Tarantino.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer></div>

  


  

  </body></html>

